// browser-use-sdk/dist/chunk-RIRVOEIU.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/core/errors.ts
var BrowserUseError = class extends Error {
  
  
  constructor(statusCode, message, detail) {
    super(message);
    this.name = "BrowserUseError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
};

// src/core/http.ts
var HttpClient = class {
  
  
  
  
  constructor(options) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.maxRetries = _nullishCoalesce(options.maxRetries, () => ( 3));
    this.timeout = _nullishCoalesce(options.timeout, () => ( 3e4));
  }
  async request(method, path, options) {
    const url = new URL(`${this.baseUrl}${path}`);
    if (_optionalChain([options, 'optionalAccess', _ => _.query])) {
      for (const [key, value] of Object.entries(options.query)) {
        if (value !== void 0 && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    const headers = {
      "X-Browser-Use-API-Key": this.apiKey
    };
    if (_optionalChain([options, 'optionalAccess', _2 => _2.body]) !== void 0) {
      headers["Content-Type"] = "application/json";
    }
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = Math.min(1e3 * 2 ** (attempt - 1), 1e4);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
      const controller = new AbortController();
      const timeoutId = _optionalChain([options, 'optionalAccess', _3 => _3.signal]) ? void 0 : setTimeout(() => controller.abort(), this.timeout);
      const signal = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _4 => _4.signal]), () => ( controller.signal));
      try {
        const response = await fetch(url.toString(), {
          method,
          headers,
          body: _optionalChain([options, 'optionalAccess', _5 => _5.body]) !== void 0 ? JSON.stringify(options.body) : void 0,
          signal
        });
        clearTimeout(timeoutId);
        if (response.ok) {
          if (response.status === 204) {
            return void 0;
          }
          return await response.json();
        }
        const shouldRetry = response.status === 429 && attempt < this.maxRetries;
        if (shouldRetry) {
          continue;
        }
        let errorBody;
        try {
          errorBody = await response.json();
        } catch (e) {
        }
        const message = typeof errorBody === "object" && errorBody !== null ? String(
          "message" in errorBody ? errorBody.message : "detail" in errorBody ? errorBody.detail : `HTTP ${response.status}`
        ) : `HTTP ${response.status}`;
        throw new BrowserUseError(response.status, message, errorBody);
      } catch (error) {
        clearTimeout(timeoutId);
        throw error;
      }
    }
    throw new Error("Unreachable: retry loop exhausted");
  }
  get(path, query) {
    return this.request("GET", path, { query });
  }
  post(path, body, query) {
    return this.request("POST", path, { body, query });
  }
  patch(path, body, query) {
    return this.request("PATCH", path, { body, query });
  }
  delete(path, query) {
    return this.request("DELETE", path, { query });
  }
};




exports.BrowserUseError = BrowserUseError; exports.HttpClient = HttpClient;
