// browser-use-sdk/dist/chunk-I75IFZAL.js
// src/core/errors.ts
var BrowserUseError = class extends Error {
  statusCode;
  detail;
  constructor(statusCode, message, detail) {
    super(message);
    this.name = "BrowserUseError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
};

// src/core/http.ts
var HttpClient = class {
  apiKey;
  baseUrl;
  maxRetries;
  timeout;
  constructor(options) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.maxRetries = options.maxRetries ?? 3;
    this.timeout = options.timeout ?? 3e4;
  }
  async request(method, path, options) {
    const url = new URL(`${this.baseUrl}${path}`);
    if (options?.query) {
      for (const [key, value] of Object.entries(options.query)) {
        if (value !== void 0 && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    const headers = {
      "X-Browser-Use-API-Key": this.apiKey
    };
    if (options?.body !== void 0) {
      headers["Content-Type"] = "application/json";
    }
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = Math.min(1e3 * 2 ** (attempt - 1), 1e4);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
      const controller = new AbortController();
      const timeoutId = options?.signal ? void 0 : setTimeout(() => controller.abort(), this.timeout);
      const signal = options?.signal ?? controller.signal;
      try {
        const response = await fetch(url.toString(), {
          method,
          headers,
          body: options?.body !== void 0 ? JSON.stringify(options.body) : void 0,
          signal
        });
        clearTimeout(timeoutId);
        if (response.ok) {
          if (response.status === 204) {
            return void 0;
          }
          return await response.json();
        }
        const shouldRetry = response.status === 429 && attempt < this.maxRetries;
        if (shouldRetry) {
          continue;
        }
        let errorBody;
        try {
          errorBody = await response.json();
        } catch {
        }
        const message = typeof errorBody === "object" && errorBody !== null ? String(
          "message" in errorBody ? errorBody.message : "detail" in errorBody ? errorBody.detail : `HTTP ${response.status}`
        ) : `HTTP ${response.status}`;
        throw new BrowserUseError(response.status, message, errorBody);
      } catch (error) {
        clearTimeout(timeoutId);
        throw error;
      }
    }
    throw new Error("Unreachable: retry loop exhausted");
  }
  get(path, query) {
    return this.request("GET", path, { query });
  }
  post(path, body, query) {
    return this.request("POST", path, { body, query });
  }
  patch(path, body, query) {
    return this.request("PATCH", path, { body, query });
  }
  delete(path, query) {
    return this.request("DELETE", path, { query });
  }
};

export {
  BrowserUseError,
  HttpClient
};
