// browser-use-sdk/dist/v3.js
import {
  BrowserUseError,
  HttpClient
} from "./chunk-I75IFZAL.js";

// src/v3/client.ts
import { z } from "zod";

// src/v3/resources/sessions.ts
var Sessions = class {
  constructor(http) {
    this.http = http;
  }
  /** Create a session and optionally dispatch a task. */
  create(body) {
    return this.http.post("/sessions", body);
  }
  /** List sessions for the authenticated project. */
  list(params) {
    return this.http.get("/sessions", params);
  }
  /** Get session details. */
  get(sessionId) {
    return this.http.get(`/sessions/${sessionId}`);
  }
  /** Stop a session or the running task. */
  stop(sessionId, body) {
    return this.http.post(`/sessions/${sessionId}/stop`, body);
  }
  /** Soft-delete a session. */
  delete(sessionId) {
    return this.http.delete(`/sessions/${sessionId}`);
  }
  /** Get presigned upload URLs for session files. */
  uploadFiles(sessionId, body) {
    return this.http.post(`/sessions/${sessionId}/files/upload`, body);
  }
  /** List files in a session's workspace. */
  files(sessionId, params) {
    return this.http.get(
      `/sessions/${sessionId}/files`,
      params
    );
  }
};

// src/v3/helpers.ts
var TERMINAL_STATUSES = /* @__PURE__ */ new Set(["idle", "stopped", "timed_out", "error"]);
var SessionRun = class {
  _createPromise;
  _sessions;
  _schema;
  _timeout;
  _interval;
  _sessionId = null;
  _result = null;
  constructor(createPromise, sessions, schema, options) {
    this._createPromise = createPromise;
    this._sessions = sessions;
    this._schema = schema;
    this._timeout = options?.timeout ?? 3e5;
    this._interval = options?.interval ?? 2e3;
  }
  /** The session ID, available after task creation resolves. */
  get sessionId() {
    return this._sessionId;
  }
  /** The full SessionResult, available after polling completes. */
  get result() {
    return this._result;
  }
  /** Enable `await client.run(...)` — polls until terminal, returns SessionResult. */
  then(onFulfilled, onRejected) {
    return this._waitForOutput().then(onFulfilled, onRejected);
  }
  async _ensureSessionId() {
    if (this._sessionId) return this._sessionId;
    const created = await this._createPromise;
    this._sessionId = created.id;
    return this._sessionId;
  }
  /** Poll session until terminal, return SessionResult. */
  async _waitForOutput() {
    const sessionId = await this._ensureSessionId();
    const deadline = Date.now() + this._timeout;
    while (Date.now() < deadline) {
      const session = await this._sessions.get(sessionId);
      if (TERMINAL_STATUSES.has(session.status)) {
        const { output, ...rest } = session;
        const parsed = this._parseOutput(output);
        this._result = { ...rest, output: parsed };
        return this._result;
      }
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await new Promise(
        (r) => setTimeout(r, Math.min(this._interval, remaining))
      );
    }
    throw new Error(
      `Session ${sessionId} did not complete within ${this._timeout}ms`
    );
  }
  _parseOutput(output) {
    if (output == null) return null;
    if (!this._schema) return output;
    const raw = typeof output === "string" ? JSON.parse(output) : output;
    return this._schema.parse(raw);
  }
};

// src/v3/client.ts
var DEFAULT_BASE_URL = "https://api.browser-use.com/api/v3";
var BrowserUse = class {
  sessions;
  http;
  constructor(options = {}) {
    const apiKey = options.apiKey ?? process.env.BROWSER_USE_API_KEY ?? "";
    if (!apiKey) {
      throw new Error(
        "No API key provided. Pass apiKey or set BROWSER_USE_API_KEY."
      );
    }
    this.http = new HttpClient({
      apiKey,
      baseUrl: options.baseUrl ?? DEFAULT_BASE_URL,
      maxRetries: options.maxRetries,
      timeout: options.timeout
    });
    this.sessions = new Sessions(this.http);
  }
  run(task, options) {
    const { schema, timeout, interval, ...rest } = options ?? {};
    const body = { task, ...rest };
    if (schema) {
      body.outputSchema = z.toJSONSchema(schema);
    }
    const promise = this.sessions.create(body);
    return new SessionRun(promise, this.sessions, schema, { timeout, interval });
  }
};
export {
  BrowserUse,
  BrowserUseError,
  SessionRun,
  Sessions
};
