// @lmnr-ai/lmnr/dist/cli.js
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/cli.ts
var cli_exports = {};
__export(cli_exports, {
  loadModule: () => loadModule
});
module.exports = __toCommonJS(cli_exports);
var import_api = require("@opentelemetry/api");
var import_argparse = require("argparse");
var esbuild = __toESM(require("esbuild"));
var glob = __toESM(require("glob"));
var import_pino = __toESM(require("pino"));
var import_pino_pretty = __toESM(require("pino-pretty"));

// package.json
var version = "0.5.2";

// src/cli.ts
var logger = (0, import_pino.default)((0, import_pino_pretty.default)({
  colorize: true,
  minimumLevel: "info"
}));
function loadModule({
  moduleText
}) {
  globalThis._evaluation = void 0;
  globalThis._set_global_evaluation = true;
  new Function("require", moduleText)(require);
  return globalThis._evaluation;
}
async function cli() {
  const [, , ...args] = process.argv;
  const parser = new import_argparse.ArgumentParser({
    prog: "lmnr",
    description: "CLI for Laminar. Use `lmnr <subcommand> --help` for more information."
  });
  parser.add_argument("-v", "--version", { action: "version", version });
  const subparsers = parser.add_subparsers({
    title: "subcommands",
    dest: "subcommand"
  });
  const parserEval = subparsers.add_parser("eval", {
    description: "Run an evaluation",
    help: "Run an evaluation"
  });
  parserEval.add_argument("file", {
    help: "A file containing the evaluation to run. If no file is provided, the evaluation will run all `*.eval.ts|js` files in the `evals` directory.",
    nargs: "?"
  });
  parserEval.add_argument("--fail-on-error", {
    help: "Fail on error. If specified, will fail if encounters a file that cannot be run",
    action: "store_true"
  });
  parserEval.set_defaults({
    func: async (args2) => {
      const files = args2.file ? [args2.file] : glob.sync("evals/**/*.eval.{ts,js}");
      files.sort();
      if (files.length === 0) {
        logger.error("No evaluation files found. Please provide a file or ensure there are files in the `evals` directory.");
        process.exit(1);
      }
      if (!args2.file) {
        logger.info(`Located ${files.length} evaluation files in evals/`);
      }
      for (const file of files) {
        logger.info(`Loading ${file}...`);
        const buildOptions = {
          entryPoints: [file],
          outfile: `tmp_out_${file}.js`,
          write: false,
          // will be loaded in memory as a temp file
          platform: "node",
          bundle: true,
          external: ["node_modules/*"]
        };
        const result = await esbuild.build(buildOptions);
        if (!result.outputFiles) {
          logger.error("Error when building: No output files found");
          if (args2.fail_on_error) {
            process.exit(1);
          }
          continue;
        }
        const outputFileText = result.outputFiles[0].text;
        const evaluation = loadModule({
          filename: args2.file,
          moduleText: outputFileText
        });
        if (!evaluation?.run) {
          logger.error(`Evaluation ${file} does not properly call evaluate()`);
          if (args2.fail_on_error) {
            process.exit(1);
          }
          continue;
        }
        await evaluation.run();
        import_api.context.disable();
        import_api.trace.disable();
        import_api.propagation.disable();
      }
    }
  });
  parser.set_defaults({
    func: () => {
      parser.print_help();
      process.exit(0);
    }
  });
  const parsed = parser.parse_args(args);
  await parsed.func(parsed);
}
cli().catch((err) => {
  logger.error(err);
  process.exit(1);
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  loadModule
});
