// @lmnr-ai/lmnr/dist/cli.mjs
#!/usr/bin/env node
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/cli.ts
import { context, propagation, trace } from "@opentelemetry/api";
import { ArgumentParser } from "argparse";
import * as esbuild from "esbuild";
import * as glob from "glob";
import pino from "pino";
import pinoPretty from "pino-pretty";

// package.json
var version = "0.5.2";

// src/cli.ts
var logger = pino(pinoPretty({
  colorize: true,
  minimumLevel: "info"
}));
function loadModule({
  moduleText
}) {
  globalThis._evaluation = void 0;
  globalThis._set_global_evaluation = true;
  new Function("require", moduleText)(__require);
  return globalThis._evaluation;
}
async function cli() {
  const [, , ...args] = process.argv;
  const parser = new ArgumentParser({
    prog: "lmnr",
    description: "CLI for Laminar. Use `lmnr <subcommand> --help` for more information."
  });
  parser.add_argument("-v", "--version", { action: "version", version });
  const subparsers = parser.add_subparsers({
    title: "subcommands",
    dest: "subcommand"
  });
  const parserEval = subparsers.add_parser("eval", {
    description: "Run an evaluation",
    help: "Run an evaluation"
  });
  parserEval.add_argument("file", {
    help: "A file containing the evaluation to run. If no file is provided, the evaluation will run all `*.eval.ts|js` files in the `evals` directory.",
    nargs: "?"
  });
  parserEval.add_argument("--fail-on-error", {
    help: "Fail on error. If specified, will fail if encounters a file that cannot be run",
    action: "store_true"
  });
  parserEval.set_defaults({
    func: async (args2) => {
      const files = args2.file ? [args2.file] : glob.sync("evals/**/*.eval.{ts,js}");
      files.sort();
      if (files.length === 0) {
        logger.error("No evaluation files found. Please provide a file or ensure there are files in the `evals` directory.");
        process.exit(1);
      }
      if (!args2.file) {
        logger.info(`Located ${files.length} evaluation files in evals/`);
      }
      for (const file of files) {
        logger.info(`Loading ${file}...`);
        const buildOptions = {
          entryPoints: [file],
          outfile: `tmp_out_${file}.js`,
          write: false,
          // will be loaded in memory as a temp file
          platform: "node",
          bundle: true,
          external: ["node_modules/*"]
        };
        const result = await esbuild.build(buildOptions);
        if (!result.outputFiles) {
          logger.error("Error when building: No output files found");
          if (args2.fail_on_error) {
            process.exit(1);
          }
          continue;
        }
        const outputFileText = result.outputFiles[0].text;
        const evaluation = loadModule({
          filename: args2.file,
          moduleText: outputFileText
        });
        if (!evaluation?.run) {
          logger.error(`Evaluation ${file} does not properly call evaluate()`);
          if (args2.fail_on_error) {
            process.exit(1);
          }
          continue;
        }
        await evaluation.run();
        context.disable();
        trace.disable();
        propagation.disable();
      }
    }
  });
  parser.set_defaults({
    func: () => {
      parser.print_help();
      process.exit(0);
    }
  });
  const parsed = parser.parse_args(args);
  await parsed.func(parsed);
}
cli().catch((err) => {
  logger.error(err);
  process.exit(1);
});
export {
  loadModule
};
