// @lmnr-ai/lmnr/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Dataset: () => EvaluationDataset,
  HumanEvaluator: () => HumanEvaluator,
  Laminar: () => Laminar,
  LaminarAttributes: () => LaminarAttributes,
  LaminarClient: () => LaminarClient,
  LaminarDataset: () => LaminarDataset,
  Span: () => import_api13.Span,
  TracingLevel: () => TracingLevel,
  evaluate: () => evaluate,
  observe: () => observe,
  withLabels: () => withLabels,
  withTracingLevel: () => withTracingLevel
});
module.exports = __toCommonJS(index_exports);

// src/client/resources/agent.ts
var import_api2 = require("@opentelemetry/api");
var import_pino2 = __toESM(require("pino"));
var import_pino_pretty2 = __toESM(require("pino-pretty"));

// src/utils.ts
var import_api = require("@opentelemetry/api");
var import_pino = __toESM(require("pino"));
var import_pino_pretty = __toESM(require("pino-pretty"));
var import_uuid = require("uuid");
var logger = (0, import_pino.default)((0, import_pino_pretty.default)({
  colorize: true,
  minimumLevel: "info"
}));
var NIL_UUID = "00000000-0000-0000-0000-000000000000";
var newUUID = () => {
  if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
    return crypto.randomUUID();
  } else {
    return (0, import_uuid.v4)();
  }
};
var otelSpanIdToUUID = (spanId) => {
  let id = spanId.toLowerCase();
  if (id.startsWith("0x")) {
    id = id.slice(2);
  }
  if (id.length !== 16) {
    logger.warn(`Span ID ${spanId} is not 16 hex chars long. This is not a valid OpenTelemetry span ID.`);
  }
  if (!/^[0-9a-f]+$/.test(id)) {
    logger.error(`Span ID ${spanId} is not a valid hex string. Generating a random UUID instead.`);
    return newUUID();
  }
  return id.padStart(32, "0").replace(/^(.{8})(.{4})(.{4})(.{4})(.{12})$/, "$1-$2-$3-$4-$5");
};
var otelTraceIdToUUID = (traceId) => {
  let id = traceId.toLowerCase();
  if (id.startsWith("0x")) {
    id = id.slice(2);
  }
  if (id.length !== 32) {
    logger.warn(`Trace ID ${traceId} is not 32 hex chars long. This is not a valid OpenTelemetry trace ID.`);
  }
  if (!/^[0-9a-f]+$/.test(id)) {
    logger.error(`Trace ID ${traceId} is not a valid hex string. Generating a random UUID instead.`);
    return newUUID();
  }
  return id.replace(/^(.{8})(.{4})(.{4})(.{4})(.{12})$/, "$1-$2-$3-$4-$5");
};
var uuidToOtelTraceId = (uuid) => uuid.replace(/-/g, "");
var uuidToOtelSpanId = (uuid) => uuid.replace(/-/g, "").slice(16);
var Semaphore = class {
  constructor(value = 1) {
    /**
     * List of promises that will be resolved when a permit becomes available.
     */
    this._waiters = [];
    if (value < 0) {
      throw new Error("Semaphore value must be >= 0");
    }
    this._value = value;
    this._waiters = [];
  }
  async acquire() {
    if (this._value > 0) {
      this._value--;
      return;
    }
    return new Promise((resolve) => {
      this._waiters.push(resolve);
    });
  }
  release() {
    if (this._waiters.length > 0) {
      const resolve = this._waiters.shift();
      resolve?.();
    } else {
      this._value++;
    }
  }
  // Python-like context manager functionality
  async using(fn) {
    try {
      await this.acquire();
      return await fn();
    } finally {
      this.release();
    }
  }
};
var tryToOtelSpanContext = (spanContext) => {
  if (typeof spanContext === "string") {
    try {
      const record = JSON.parse(spanContext);
      return recordToOtelSpanContext(record);
    } catch (e) {
      throw new Error(`Failed to parse span context ${spanContext}. The string must be a json representation of a LaminarSpanContext.Error: ${e instanceof Error ? e.message : String(e)}`);
    }
  } else if (isRecord(spanContext)) {
    return recordToOtelSpanContext(spanContext);
  } else if (typeof spanContext.traceId === "string" && typeof spanContext.spanId === "string" && spanContext.traceId.length === 32 && spanContext.spanId.length === 16) {
    logger.warn("The span context is already an OpenTelemetry SpanContext. Returning it as is. Please use `LaminarSpanContext` objects instead.");
    return spanContext;
  } else {
    throw new Error(`Invalid span context ${JSON.stringify(spanContext)}. Must be a LaminarSpanContext or its json representation.`);
  }
};
var recordToOtelSpanContext = (record) => {
  if (typeof record.spanId === "string" && typeof record.traceId === "string") {
    return {
      spanId: uuidToOtelSpanId(record?.spanId ?? record?.["span_id"]),
      traceId: uuidToOtelTraceId(record?.traceId ?? record?.["trace_id"]),
      isRemote: record?.isRemote ?? record?.["is_remote"] ?? false,
      traceFlags: record?.traceFlags ?? import_api.TraceFlags.SAMPLED
    };
  } else {
    throw new Error(`Invalid span context ${JSON.stringify(record)}. Must be a json representation of a LaminarSpanContext.`);
  }
};
var isRecord = (value) => typeof value === "object" && !Array.isArray(value) && value !== null;

// src/client/resources/index.ts
var BaseResource = class {
  constructor(baseHttpUrl, projectApiKey) {
    this.baseHttpUrl = baseHttpUrl;
    this.projectApiKey = projectApiKey;
  }
  headers() {
    return {
      Authorization: `Bearer ${this.projectApiKey}`,
      "Content-Type": "application/json",
      Accept: "application/json"
    };
  }
  async handleError(response) {
    const errorMsg = await response.text();
    throw new Error(`${response.status} ${errorMsg}`);
  }
};

// src/client/resources/agent.ts
var logger2 = (0, import_pino2.default)((0, import_pino_pretty2.default)({
  colorize: true,
  minimumLevel: "info"
}));
var AgentResource = class extends BaseResource {
  constructor(baseHttpUrl, projectApiKey) {
    super(baseHttpUrl, projectApiKey);
  }
  /**
   * Run Laminar index agent
   *
   * @param { RunAgentOptions } options - The options for running the agent
   * @param { string } options.prompt - The prompt for the agent
   * @param { string } [options.parentSpanContext] - The parent span context for tracing
   * @param { ModelProvider } [options.modelProvider] - LLM provider to use
   * @param { string } [options.model] - The model name as specified in the provider API
   * @param { boolean } [options.stream] - Whether to stream the response. Defaults to false.
   * @param { boolean } [options.enableThinking] - Whether to enable thinking in the underlying
   * LLM. Defaults to true.
   * @param { number } [options.timeout] - The timeout in seconds for the agent.
   * Note: This is a soft timeout. The agent will finish a step even after the timeout has
   * been reached.
   * @param { string } [options.cdpUrl] - The URL of an existing Chrome DevTools Protocol
   * (CDP) browser instance.
   * @param { number } [options.maxSteps] - The maximum number of steps the agent can take.
   * Defaults to 100.
   * @param { number } [options.thinkingTokenBudget] - The maximum number of tokens the underlying
   * LLM can spend on thinking per step, if supported by the LLM provider.
   * @param { string } [options.startUrl] - The URL to start the agent on.
   * Make sure it's a valid URL - refer to https://playwright.dev/docs/api/class-page#page-goto
   * If not specified, the agent will infer this from the prompt.
   * @param { boolean } [options.returnScreenshots] - Whether to return screenshots with
   * each step. Defaults to false.
   * @param { boolean } [options.returnAgentState] - Whether to return the agent state.
   * Agent state can be used to resume the agent in subsequent runs.
   * CAUTION: Agent state is a very large object. Defaults to false.
   * @param { boolean } [options.returnStorageState] - Whether to return the storage state.
   * Storage state includes browser cookies, auth, etc.
   * CAUTION: Storage state is a relatively large object. Defaults to false.
   * @returns { Promise<AgentOutput | ReadableStream<RunAgentResponseChunk>> }
   *    The agent output or a stream of response chunks
   */
  async run({
    prompt,
    parentSpanContext,
    modelProvider,
    model,
    stream,
    enableThinking,
    timeout,
    cdpUrl,
    agentState,
    storageState,
    maxSteps,
    thinkingTokenBudget,
    startUrl,
    returnScreenshots,
    returnAgentState,
    returnStorageState
  }) {
    let requestParentSpanContext = parentSpanContext;
    if (!requestParentSpanContext) {
      const currentSpan = import_api2.trace.getActiveSpan();
      if (currentSpan && currentSpan.isRecording()) {
        const traceId = otelTraceIdToUUID(currentSpan.spanContext().traceId);
        const spanId = otelSpanIdToUUID(currentSpan.spanContext().spanId);
        requestParentSpanContext = JSON.stringify({
          trace_id: traceId,
          span_id: spanId,
          is_remote: currentSpan.spanContext().isRemote
        });
      }
    }
    const request = {
      prompt,
      parentSpanContext: requestParentSpanContext,
      modelProvider,
      model,
      stream: true,
      enableThinking: enableThinking ?? true,
      timeout,
      cdpUrl,
      agentState,
      storageState,
      maxSteps,
      thinkingTokenBudget,
      startUrl,
      returnScreenshots: returnScreenshots ?? false,
      returnAgentState: returnAgentState ?? false,
      returnStorageState: returnStorageState ?? false
    };
    if (stream) {
      return this.runStreaming(request);
    } else {
      return await this.runNonStreaming(request);
    }
  }
  /**
   * Run agent in streaming mode
   */
  async runStreaming(request) {
    const response = await fetch(this.baseHttpUrl + "/v1/agent/run", {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(request)
    });
    if (!response.ok) {
      await this.handleError(response);
    }
    if (!response.body) {
      throw new Error("Response body is null");
    }
    return response.body.pipeThrough(new TextDecoderStream()).pipeThrough(new TransformStream({
      start() {
        this.buffer = "";
      },
      transform(chunk, controller) {
        this.buffer += chunk;
        const lines = this.buffer.split("\n");
        this.buffer = lines.pop() || "";
        for (const line of lines) {
          if (line.startsWith("[DONE]")) {
            controller.terminate();
            return;
          }
          if (!line.startsWith("data: ")) {
            continue;
          }
          const jsonStr = line.substring(6);
          if (jsonStr) {
            try {
              const parsed = JSON.parse(jsonStr);
              controller.enqueue(parsed);
            } catch (error) {
              logger2.error(`Error parsing JSON: ${error instanceof Error ? error.message : String(error)}`);
            }
          }
        }
      },
      flush(controller) {
        if (this.buffer) {
          if (this.buffer.startsWith("data: ")) {
            const jsonStr = this.buffer.substring(6);
            if (jsonStr) {
              try {
                const parsed = JSON.parse(jsonStr);
                controller.enqueue(parsed);
              } catch (error) {
                logger2.error(`Error parsing JSON: ${error instanceof Error ? error.message : String(error)}`);
              }
            }
          }
        }
      }
    }));
  }
  /**
   * Run agent in non-streaming mode
   */
  async runNonStreaming(request) {
    const stream = await this.runStreaming(request);
    const reader = stream.getReader();
    let finalChunk = null;
    let errorChunk = null;
    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        if (value.chunkType === "finalOutput") {
          finalChunk = value;
          break;
        } else if (value.chunkType === "error") {
          errorChunk = value;
          break;
        } else if (value.chunkType === "timeout") {
          errorChunk = {
            chunkType: "error",
            messageId: value.messageId,
            error: "Timeout"
          };
          break;
        }
      }
    } finally {
      reader.releaseLock();
    }
    if (errorChunk) {
      throw new Error(errorChunk.error);
    }
    return finalChunk?.content || { result: { isDone: true } };
  }
};

// package.json
var version = "0.5.2";

// src/version.ts
var getLangVersion = () => {
  if (process?.versions?.node) {
    return `node@${process.versions.node}`;
  }
};

// src/client/resources/browser-events.ts
var BrowserEventsResource = class extends BaseResource {
  constructor(baseHttpUrl, projectApiKey) {
    super(baseHttpUrl, projectApiKey);
  }
  async send({
    sessionId,
    traceId,
    events
  }) {
    const payload = {
      sessionId,
      traceId,
      events,
      source: getLangVersion() ?? "javascript",
      sdkVersion: version
    };
    const jsonString = JSON.stringify(payload);
    const blob = new Blob([jsonString], { type: "application/json" });
    const compressedStream = blob.stream().pipeThrough(new CompressionStream("gzip"));
    const compressedResponse = new Response(compressedStream);
    const compressedData = await compressedResponse.arrayBuffer();
    const response = await fetch(this.baseHttpUrl + "/v1/events", {
      method: "POST",
      headers: {
        ...this.headers(),
        "Content-Encoding": "gzip"
      },
      body: compressedData
    });
    if (!response.ok) {
      await this.handleError(response);
    }
  }
};

// src/client/resources/evals.ts
var EvalsResource = class extends BaseResource {
  constructor(baseHttpUrl, projectApiKey) {
    super(baseHttpUrl, projectApiKey);
  }
  /**
   * Initialize an evaluation.
   *
   * @param {string} name - Name of the evaluation
   * @param {string} groupName - Group name of the evaluation
   * @returns {Promise<InitEvaluationResponse>} Response from the evaluation initialization
   */
  async init(name, groupName) {
    const response = await fetch(this.baseHttpUrl + "/v1/evals", {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ name: name ?? null, groupName: groupName ?? null })
    });
    if (!response.ok) {
      await this.handleError(response);
    }
    return response.json();
  }
  /**
   * Save evaluation datapoints.
   *
   * @param {Object} options - Save datapoints options
   * @param {string} options.evalId - ID of the evaluation
   * @param {EvaluationDatapoint<D, T, O>[]} options.datapoints - Datapoint to add
   * @param {string} [options.groupName] - Group name of the evaluation
   * @returns {Promise<void>} Response from the datapoint addition
   */
  async saveDatapoints({
    evalId,
    datapoints,
    groupName
  }) {
    const response = await fetch(this.baseHttpUrl + `/v1/evals/${evalId}/datapoints`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ points: datapoints, groupName: groupName ?? null })
    });
    if (!response.ok) {
      await this.handleError(response);
    }
  }
  /**
   * Get evaluation datapoints.
   *
   * @param {Object} options - Get datapoints options
   * @param {string} options.datasetName - Name of the dataset
   * @param {number} options.offset - Offset at which to start the query
   * @param {number} options.limit - Maximum number of datapoints to return
   * @returns {Promise<GetDatapointsResponse>} Response from the datapoint retrieval
   */
  async getDatapoints({
    datasetName,
    offset,
    limit
  }) {
    const params = new URLSearchParams({
      name: datasetName,
      offset: offset.toString(),
      limit: limit.toString()
    });
    const response = await fetch(
      this.baseHttpUrl + `/v1/datasets/datapoints?${params.toString()}`,
      {
        method: "GET",
        headers: this.headers()
      }
    );
    if (!response.ok) {
      await this.handleError(response);
    }
    return await response.json();
  }
};

// src/client/index.ts
var LaminarClient = class {
  constructor({
    baseUrl,
    projectApiKey,
    port
  }) {
    this.projectApiKey = projectApiKey ?? process.env.LMNR_PROJECT_API_KEY;
    const httpPort = port ?? (baseUrl?.match(/:\d{1,5}$/g) ? parseInt(baseUrl.match(/:\d{1,5}$/g)[0].slice(1)) : 443);
    this.baseUrl = `${baseUrl?.replace(/\/$/, "").replace(/:\d{1,5}$/g, "") ?? "https://api.lmnr.ai"}:${httpPort}`;
    this._agent = new AgentResource(this.baseUrl, this.projectApiKey);
    this._browserEvents = new BrowserEventsResource(this.baseUrl, this.projectApiKey);
    this._evals = new EvalsResource(this.baseUrl, this.projectApiKey);
  }
  get agent() {
    return this._agent;
  }
  get browserEvents() {
    return this._browserEvents;
  }
  get evals() {
    return this._evals;
  }
};

// src/datasets.ts
var DEFAULT_FETCH_SIZE = 25;
var EvaluationDataset = class {
  async slice(start, end) {
    const result = [];
    for (let i = Math.max(start, 0); i < Math.min(end, await this.size()); i++) {
      result.push(await this.get(i));
    }
    return result;
  }
};
var LaminarDataset = class extends EvaluationDataset {
  constructor(name, fetchSize) {
    super();
    this.fetchedItems = [];
    this.len = null;
    this.offset = 0;
    this.client = void 0;
    this.name = name;
    this.fetchSize = fetchSize || DEFAULT_FETCH_SIZE;
  }
  setClient(client) {
    this.client = client;
  }
  async fetchBatch() {
    if (!this.client) {
      throw new Error("Client not set");
    }
    const resp = await this.client.evals.getDatapoints({
      datasetName: this.name,
      offset: this.offset,
      limit: this.fetchSize
    });
    this.fetchedItems = this.fetchedItems.concat(resp.items);
    this.offset = this.fetchedItems.length;
    if (this.len === null) {
      this.len = resp.totalCount;
    }
  }
  async size() {
    if (this.len === null) {
      await this.fetchBatch();
    }
    return this.len;
  }
  async get(index) {
    if (index >= this.fetchedItems.length) {
      await this.fetchBatch();
    }
    return this.fetchedItems[index];
  }
};

// src/decorators.ts
var import_api11 = require("@opentelemetry/api");

// src/sdk/configuration/index.ts
var import_api9 = require("@opentelemetry/api");

// src/sdk/errors/index.ts
var SEVERITY = {
  Warning: "Warning",
  Error: "Error",
  Critical: "Critical"
};
var TraceloopError = class extends Error {
  constructor(message, severity = SEVERITY.Error) {
    super(message);
    this.severity = severity;
  }
};
var InitializationError = class extends TraceloopError {
  constructor(message, cause) {
    super(message ?? "Failed to initialize Traceloop SDK", SEVERITY.Critical);
    this.underlyingCause = cause;
  }
};

// src/sdk/tracing/index.ts
var import_api8 = require("@opentelemetry/api");
var import_core = require("@opentelemetry/core");
var import_exporter_trace_otlp_proto = require("@opentelemetry/exporter-trace-otlp-proto");
var import_instrumentation4 = require("@opentelemetry/instrumentation");
var import_sdk_trace_node = require("@opentelemetry/sdk-trace-node");
var import_instrumentation_anthropic = require("@traceloop/instrumentation-anthropic");
var import_instrumentation_azure = require("@traceloop/instrumentation-azure");
var import_instrumentation_bedrock = require("@traceloop/instrumentation-bedrock");
var import_instrumentation_chromadb = require("@traceloop/instrumentation-chromadb");
var import_instrumentation_cohere = require("@traceloop/instrumentation-cohere");
var import_instrumentation_langchain = require("@traceloop/instrumentation-langchain");
var import_instrumentation_llamaindex = require("@traceloop/instrumentation-llamaindex");
var import_instrumentation_openai = require("@traceloop/instrumentation-openai");
var import_instrumentation_pinecone = require("@traceloop/instrumentation-pinecone");
var import_instrumentation_qdrant = require("@traceloop/instrumentation-qdrant");
var import_instrumentation_together = require("@traceloop/instrumentation-together");
var import_instrumentation_vertexai = require("@traceloop/instrumentation-vertexai");
var import_pino7 = __toESM(require("pino"));
var import_pino_pretty7 = __toESM(require("pino-pretty"));

// src/browser/playwright.ts
var import_api5 = require("@opentelemetry/api");
var import_instrumentation = require("@opentelemetry/instrumentation");
var import_fs = require("fs");
var import_promises = require("fs/promises");
var import_path2 = __toESM(require("path"));
var import_pino5 = __toESM(require("pino"));
var import_pino_pretty5 = __toESM(require("pino-pretty"));

// src/laminar.ts
var import_grpc_js = require("@grpc/grpc-js");
var import_api4 = require("@opentelemetry/api");
var import_exporter_trace_otlp_grpc = require("@opentelemetry/exporter-trace-otlp-grpc");
var import_pino3 = __toESM(require("pino"));
var import_pino_pretty3 = __toESM(require("pino-pretty"));

// src/sdk/tracing/attributes.ts
var import_ai_semantic_conventions = require("@traceloop/ai-semantic-conventions");
var SPAN_INPUT = "lmnr.span.input";
var SPAN_OUTPUT = "lmnr.span.output";
var SPAN_TYPE = "lmnr.span.type";
var SPAN_PATH = "lmnr.span.path";
var SPAN_IDS_PATH = "lmnr.span.ids_path";
var SPAN_INSTRUMENTATION_SOURCE = "lmnr.span.instrumentation_source";
var SPAN_SDK_VERSION = "lmnr.span.sdk_version";
var SPAN_LANGUAGE_VERSION = "lmnr.span.language_version";
var OVERRIDE_PARENT_SPAN = "lmnr.internal.override_parent_span";
var TRACE_HAS_BROWSER_SESSION = "lmnr.internal.has_browser_session";
var EXTRACTED_FROM_NEXT_JS = "lmnr.span.extracted_from.next_js";
var ASSOCIATION_PROPERTIES = "lmnr.association.properties";
var SESSION_ID = "lmnr.association.properties.session_id";
var ASSOCIATION_PROPERTIES_OVERRIDES = {
  "span_type": SPAN_TYPE
};
var LaminarAttributes = {
  // == This is the minimum set of attributes for a proper LLM span ==
  //
  // not SpanAttributes.LLM_USAGE_PROMPT_TOKENS
  INPUT_TOKEN_COUNT: "gen_ai.usage.input_tokens",
  // not SpanAttributes.LLM_USAGE_COMPLETION_TOKENS
  OUTPUT_TOKEN_COUNT: "gen_ai.usage.output_tokens",
  TOTAL_TOKEN_COUNT: import_ai_semantic_conventions.SpanAttributes.LLM_USAGE_TOTAL_TOKENS,
  PROVIDER: import_ai_semantic_conventions.SpanAttributes.LLM_SYSTEM,
  REQUEST_MODEL: import_ai_semantic_conventions.SpanAttributes.LLM_REQUEST_MODEL,
  RESPONSE_MODEL: import_ai_semantic_conventions.SpanAttributes.LLM_RESPONSE_MODEL,
  //
  // == End of minimum set ==
  // == Additional attributes ==
  //
  INPUT_COST: "gen_ai.usage.input_cost",
  OUTPUT_COST: "gen_ai.usage.output_cost",
  TOTAL_COST: "gen_ai.usage.cost"
  //
  // == End of additional attributes ==
};

// src/sdk/tracing/tracing.ts
var import_api3 = require("@opentelemetry/api");
var TRACER_NAME = "lmnr.tracer";
var SPAN_PATH_KEY = (0, import_api3.createContextKey)("span_path");
var ASSOCIATION_PROPERTIES_KEY = (0, import_api3.createContextKey)(
  "association_properties"
);
var getTracer = () => import_api3.trace.getTracer(TRACER_NAME);
var getSpanPath = (entityContext) => {
  const path4 = entityContext.getValue(SPAN_PATH_KEY);
  return path4 ? `${path4}` : void 0;
};

// src/laminar.ts
var logger3 = (0, import_pino3.default)((0, import_pino_pretty3.default)({
  colorize: true,
  minimumLevel: "info"
}));
var Laminar = class {
  /**
   * Initialize Laminar context across the application.
   * This method must be called before using any other Laminar methods or decorators.
   *
   * @param project_api_key - Laminar project api key. You can generate one by going
   * to the projects settings page on the Laminar dashboard.
   * If not specified, it will try to read from the LMNR_PROJECT_API_KEY environment variable.
   * @param env - Default environment passed to `run` requests, unless overridden at request
   * time. Usually, model provider keys are stored here.
   * @param baseUrl - Laminar API url. Do not include the port, use
   * `httpPort` and `grpcPort` instead.
   * If not specified, defaults to https://api.lmnr.ai.
   * @param httpPort - Laminar API http port.
   * If not specified, defaults to 443.
   * @param grpcPort - Laminar API grpc port.
   * If not specified, defaults to 8443.
   * @param instrumentModules - List of modules to instrument.
   * If not specified, all auto-instrumentable modules will be instrumented, which include
   * LLM calls (OpenAI, Anthropic, etc), Langchain, VectorDB calls (Pinecone, Qdrant, etc).
   * Pass an empty object {} to disable any kind of automatic instrumentation.
   * If you only want to auto-instrument specific modules, then pass them in the object.
   * @param preserveNextJsSpans - Whether to preserve thorough/verbose spans created by Next.js.
   * By default, Laminar ignores the Next.js spans, but keeps spans nested within them.
   * @param disableBatch - Whether to disable batching of spans. Useful for debug
   * environments. If true, spans will be sent immediately using {@link SimpleSpanProcessor}
   * instead of {@link BatchSpanProcessor}.
   * @param traceExportTimeoutMillis - Timeout for trace export. Defaults to 30_000 (30 seconds),
   * which is over the default OTLP exporter timeout of 10_000 (10 seconds).
   * @param useExternalTracerProvider - [ADVANCED] Only use if you are using another
   * node-based tracer provider. Defaults to false.
   * If `true`, the SDK will not initialize its own tracer provider. Be very careful.
   * If you set this to `true`, but the external provider does not extend/implement
   * {@link BasicTracerProvider} or is not a {@link ProxyTracerProvider} that internally
   * delegates to a {@link BasicTracerProvider}, the SDK will not function correctly.
   * The only setting this has been tested with is `@opentelemetry/auto-instrumentations-node`.
   * {@link https://opentelemetry.io/docs/zero-code/js/#configuring-the-module} - this
   * initializes a {@link ProxyTracerProvider} internally which delegates to a
   * {@link NodeTracerProvider}, which in turn, extends {@link BasicTracerProvider}.
   *
   * @example
   * import { Laminar as L } from '@lmnr-ai/lmnr';
   * import { OpenAI } from 'openai';
   * import * as ChainsModule from "langchain/chains";
   *
   * // Initialize Laminar while auto-instrumenting Langchain and OpenAI modules.
   * L.initialize({ projectApiKey: "<LMNR_PROJECT_API_KEY>", instrumentModules: {
   *  langchain: {
   *      chainsModule: ChainsModule
   *  },
   *  openAI: OpenAI
   * } });
   *
   * @throws {Error} - If project API key is not set
   */
  static initialize({
    projectApiKey,
    env,
    baseUrl,
    httpPort,
    grpcPort,
    instrumentModules,
    useExternalTracerProvider,
    preserveNextJsSpans,
    disableBatch,
    traceExportTimeoutMillis,
    logLevel,
    maxExportBatchSize
  }) {
    const key = projectApiKey ?? process?.env?.LMNR_PROJECT_API_KEY;
    if (key === void 0) {
      throw new Error(
        "Please initialize the Laminar object with your project API key or set the LMNR_PROJECT_API_KEY environment variable"
      );
    }
    this.projectApiKey = key;
    const url = baseUrl ?? process?.env?.LMNR_BASE_URL ?? "https://api.lmnr.ai";
    const port = httpPort ?? (url.match(/:\d{1,5}$/g) ? parseInt(url.match(/:\d{1,5}$/g)[0].slice(1)) : 443);
    const urlWithoutSlash = url.replace(/\/$/, "").replace(/:\d{1,5}$/g, "");
    this.baseHttpUrl = `${urlWithoutSlash}:${port}`;
    this.baseGrpcUrl = `${urlWithoutSlash}:${grpcPort ?? 8443}`;
    this.isInitialized = true;
    this.env = env ?? {};
    const metadata = new import_grpc_js.Metadata();
    metadata.set("authorization", `Bearer ${this.projectApiKey}`);
    const exporter = new import_exporter_trace_otlp_grpc.OTLPTraceExporter({
      url: `${this.baseGrpcUrl}`,
      metadata,
      // default is 10 seconds, increase to 30 seconds
      timeoutMillis: traceExportTimeoutMillis ?? 3e4
    });
    initializeTracing({
      baseUrl: this.baseHttpUrl,
      apiKey: this.projectApiKey,
      exporter,
      silenceInitializationMessage: true,
      instrumentModules,
      disableBatch,
      useExternalTracerProvider,
      preserveNextJsSpans,
      logLevel: logLevel ?? "error",
      maxExportBatchSize
    });
  }
  /**
   * Check if Laminar has been initialized. Utility to make sure other methods
   * are called after initialization.
   */
  static initialized() {
    return this.isInitialized;
  }
  /**
   * Sets the environment that will be sent to Laminar requests.
   *
   * @param env - The environment variables to override. If not provided, the
   * current environment will not be modified.
   */
  static setEnv(env) {
    if (env) {
      this.env = env;
    }
  }
  /**
   * Sets the project API key for authentication with Laminar.
   *
   * @param projectApiKey - The API key to be set. If not provided, the existing
   * API key will not be modified.
   */
  static setProjectApiKey(projectApiKey) {
    if (projectApiKey) {
      this.projectApiKey = projectApiKey;
    }
  }
  /**
   * Associates an event with the current span. If event with such name never
   * existed, Laminar will create a new event and infer its type from the value.
   * If the event already exists, Laminar will append the value to the event
   * if and only if the value is of a matching type. Otherwise, the event won't
   * be recorded. Supported types are string, number, and boolean. If the value
   * is `null`, event is considered a boolean tag with the value of `true`.
   *
   * @param name - The name of the event.
   * @param value - The value of the event. Must be a primitive type. If not
   * specified, boolean true is assumed in the backend.
   * @param timestamp - The timestamp of the event. If not specified, relies on
   * the underlying OpenTelemetry implementation.
   * If specified as an integer, it must be epoch nanoseconds.
   */
  static event(name, value, timestamp) {
    const currentSpan = import_api4.trace.getActiveSpan();
    if (currentSpan === void 0 || !(0, import_api4.isSpanContextValid)(currentSpan.spanContext())) {
      logger3.warn(
        `\`Laminar().event()\` called outside of span context. Event '${name}' will not be recorded in the trace. Make sure to annotate the function with a decorator`
      );
      return;
    }
    const event = {
      "lmnr.event.type": "default"
    };
    if (value !== void 0) {
      event["lmnr.event.value"] = value;
    }
    currentSpan.addEvent(name, event, timestamp);
  }
  /**
   * Sets the session information for the current span and returns the
   * context to use for the following spans. Returns the result of the
   * function execution, so can be used in an `await` statement.
   *
   * @param sessionId - The session ID to associate with the context.
   * @param fn - Function to execute within the session context.
   * @returns The result of the function execution.
   *
   * @example
   * import { Laminar, observe } from '@lmnr-ai/lmnr';
   * const result = await Laminar.withSession("session1234", async () => {
   *   // Your code here
   * });
   */
  static withSession(sessionId, fn) {
    const currentSpan = import_api4.trace.getActiveSpan();
    if (currentSpan !== void 0 && (0, import_api4.isSpanContextValid)(currentSpan.spanContext())) {
      if (sessionId) {
        currentSpan.setAttribute(SESSION_ID, sessionId);
      }
    }
    let associationProperties = {};
    if (sessionId) {
      associationProperties = { ...associationProperties, "session_id": sessionId };
    }
    let entityContext = import_api4.context.active();
    const currentAssociationProperties = entityContext.getValue(ASSOCIATION_PROPERTIES_KEY);
    if (associationProperties && Object.keys(associationProperties).length > 0) {
      entityContext = entityContext.setValue(
        ASSOCIATION_PROPERTIES_KEY,
        { ...currentAssociationProperties ?? {}, ...associationProperties }
      );
    }
    return import_api4.context.with(entityContext, fn);
  }
  /**
   * Sets the metadata for the current span and returns the context to use for
   * the following spans. Returns the result of the function execution, so can
   * be used in an `await` statement.
   *
   * @param metadata - The metadata to associate with the context. Set of string key
   * string value pairs.
   * @param fn - Function to execute within the metadata context.
   * @returns The result of the function execution.
   *
   * @example
   * import { Laminar } from '@lmnr-ai/lmnr';
   * const result = await Laminar.withMetadata(
   *   {
   *     "my_metadata_key": "my_metadata_value"
   *   },
   *   async () => {
   *     // Your code here
   *   }
   * );
   */
  static withMetadata(metadata, fn) {
    const currentSpan = import_api4.trace.getActiveSpan();
    if (currentSpan !== void 0 && (0, import_api4.isSpanContextValid)(currentSpan.spanContext())) {
      for (const [key, value] of Object.entries(metadata)) {
        currentSpan.setAttribute(`${ASSOCIATION_PROPERTIES}.metadata.${key}`, value);
      }
    }
    let metadataAttributes = {};
    for (const [key, value] of Object.entries(metadata)) {
      metadataAttributes = { ...metadataAttributes, [`metadata.${key}`]: value };
    }
    let entityContext = import_api4.context.active();
    const currentAssociationProperties = entityContext.getValue(ASSOCIATION_PROPERTIES_KEY);
    if (metadataAttributes && Object.keys(metadataAttributes).length > 0) {
      entityContext = entityContext.setValue(
        ASSOCIATION_PROPERTIES_KEY,
        { ...currentAssociationProperties ?? {}, ...metadataAttributes }
      );
    }
    return import_api4.context.with(entityContext, fn);
  }
  /**
   * Set attributes for the current span. Useful for manual
   * instrumentation.
   * @param attributes - The attributes to set for the current span.
   *
   * @example
   * import { Laminar as L, observe } from '@lmnr-ai/laminar';
   * await observe({ name: 'mySpanName', spanType: 'LLM' }, async (msg: string) => {
   *   const response = await myCustomCallToOpenAI(msg);
   *   L.setSpanAttributes({
   *     [LaminarAttributes.PROVIDER]: 'openai',
   *     [LaminarAttributes.REQUEST_MODEL]: "requested_model",
   *     [LaminarAttributes.RESPONSE_MODEL]: response.model,
   *     [LaminarAttributes.INPUT_TOKEN_COUNT]: response.usage.prompt_tokens,
   *     [LaminarAttributes.OUTPUT_TOKEN_COUNT]: response.usage.completion_tokens,
   *   })
   * }, userMessage);
   */
  static setSpanAttributes(attributes) {
    const currentSpan = import_api4.trace.getActiveSpan();
    if (currentSpan !== void 0 && (0, import_api4.isSpanContextValid)(currentSpan.spanContext())) {
      for (const [key, value] of Object.entries(attributes)) {
        currentSpan.setAttribute(key, value);
      }
    }
  }
  /**
   * Set the output of the current span. Useful for manual instrumentation.
   * @param output - Output of the span. Will be sent as an attribute, so must
   * be serializable to JSON.
   */
  static setSpanOutput(output) {
    if (output == null) {
      return;
    }
    const currentSpan = import_api4.trace.getActiveSpan();
    if (currentSpan !== void 0 && (0, import_api4.isSpanContextValid)(currentSpan.spanContext())) {
      currentSpan.setAttribute(SPAN_OUTPUT, JSON.stringify(output));
    }
  }
  /**
   * Start a new span, but don't set it as active. Useful for
   * manual instrumentation. If span type is 'LLM', you should report usage
   * manually. See {@link setSpanAttributes} for more information.
   *
   * @param name - name of the span
   * @param input - input to the span. Will be sent as an attribute, so must
   * be JSON serializable
   * @param span_type - type of the span. Defaults to 'DEFAULT'
   * @param context - raw OpenTelemetry context to bind the span to.
   * @param traceId - [EXPERIMENTAL] override the trace id for the span. If not
   * provided, uses the current trace id.
   * @returns The started span.
   *
   * @example
   * import { Laminar, observe } from '@lmnr-ai/lmnr';
   * const foo = async (span: Span) => {
   *   await Laminar.withSpan(span, async () => {
   *     await observe({ name: 'foo' }, async () => {
   *       // Your code here
   *     })
   *   })
   * };
   * const bar = async (span: Span) => {
   *   await Laminar.withSpan(span, async () => {
   *     await openai_client.chat.completions.create();
   *   })
   * };
   *
   * const parentSpan = Laminar.startSpan({name: "outer"});
   * foo(parentSpan);
   * await bar(parentSpan);
   * // IMPORTANT: Don't forget to end the span!
   * parentSpan.end();
   *
   * // Results in:
   * // | outer
   * // |  | foo
   * // |  |  | ...
   * // |  | bar
   * // |  |  | openai.chat
   */
  static startSpan({
    name,
    input,
    spanType,
    context: context4,
    parentSpanContext,
    labels
  }) {
    let entityContext = context4 ?? import_api4.context.active();
    if (parentSpanContext) {
      const spanContext = tryToOtelSpanContext(parentSpanContext);
      entityContext = import_api4.trace.setSpan(entityContext, import_api4.trace.wrapSpanContext(spanContext));
    }
    const labelProperties = labels ? { [`${ASSOCIATION_PROPERTIES}.labels`]: labels } : {};
    const attributes = {
      [SPAN_TYPE]: spanType ?? "DEFAULT",
      ...labelProperties
    };
    const span = getTracer().startSpan(name, { attributes }, entityContext);
    if (input) {
      span.setAttribute(SPAN_INPUT, JSON.stringify(input));
    }
    return span;
  }
  /**
   * A utility wrapper around OpenTelemetry's `context.with()`. Useful for
   * passing spans around in manual instrumentation:
   *
   * @param span - Parent span to bind the execution to.
   * @param fn - Function to execute within the span context.
   * @param endOnExit - Whether to end the span after the function has
   * executed. Defaults to `false`. If `false`, you MUST manually call
   * `span.end()` at the end of the execution, so that spans are not lost.
   * @returns The result of the function execution.
   *
   * See {@link startSpan} docs for a usage example
   */
  static withSpan(span, fn, endOnExit) {
    return import_api4.context.with(import_api4.trace.setSpan(import_api4.context.active(), span), () => {
      try {
        const result = fn();
        if (result instanceof Promise) {
          return result.finally(() => {
            if (endOnExit !== void 0 && endOnExit) {
              span.end();
            }
          });
        }
        if (endOnExit !== void 0 && endOnExit) {
          span.end();
        }
        return result;
      } catch (error) {
        span.recordException(error);
        if (endOnExit !== void 0 && endOnExit) {
          span.end();
        }
        throw error;
      }
    });
  }
  static serializeLaminarSpanContext(span) {
    const laminarSpanContext = this.getLaminarSpanContext(span);
    if (laminarSpanContext === null) {
      return null;
    }
    return JSON.stringify(laminarSpanContext);
  }
  static getLaminarSpanContext(span) {
    const currentSpan = span ?? import_api4.trace.getActiveSpan();
    if (currentSpan === void 0 || !(0, import_api4.isSpanContextValid)(currentSpan.spanContext())) {
      return null;
    }
    return {
      traceId: otelTraceIdToUUID(currentSpan.spanContext().traceId),
      spanId: otelSpanIdToUUID(currentSpan.spanContext().spanId),
      isRemote: currentSpan.spanContext().isRemote ?? false
    };
  }
  static async shutdown() {
    await forceFlush();
  }
  static getHttpUrl() {
    return this.baseHttpUrl;
  }
  static getProjectApiKey() {
    return this.projectApiKey;
  }
};
Laminar.env = {};
Laminar.isInitialized = false;

// src/browser/utils.ts
var import_path = __toESM(require("path"));
var import_pino4 = __toESM(require("pino"));
var import_pino_pretty4 = __toESM(require("pino-pretty"));
var import_url = require("url");
var import_meta = {};
var logger4 = (0, import_pino4.default)((0, import_pino_pretty4.default)({
  colorize: true,
  minimumLevel: "info"
}));
var getDirname = () => {
  if (typeof __dirname !== "undefined") {
    return __dirname;
  }
  if (typeof import_meta?.url !== "undefined") {
    return import_path.default.dirname((0, import_url.fileURLToPath)(import_meta.url));
  }
  return process.cwd();
};
var collectAndSendPageEvents = async (client, page, sessionId, traceId) => {
  try {
    if (page.isClosed()) {
      return;
    }
    const hasFunction = await page.evaluate(
      () => typeof window.lmnrGetAndClearEvents === "function"
    );
    if (!hasFunction) {
      return;
    }
    const events = await page.evaluate(
      async () => await window.lmnrGetAndClearEvents()
    );
    if (events == null || events.length === 0) {
      return;
    }
    await client.browserEvents.send({
      sessionId,
      traceId,
      events
    });
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    if (message.includes("Execution context was destroyed")) {
      logger4.info(`Tried to flush events from a closed page. Continuing...`);
    } else {
      logger4.error(`Error sending events: ${message}`);
    }
  }
};
var cleanStagehandLLMClient = (llmClient) => Object.fromEntries(
  Object.entries(llmClient).filter(([key]) => key !== "client").map(
    ([key, value]) => key === "clientOptions" ? [
      key,
      Object.fromEntries(
        Object.entries(value).filter(([key2]) => key2 !== "apiKey")
      )
    ] : [key, value]
  )
);

// src/browser/playwright.ts
var logger5 = (0, import_pino5.default)((0, import_pino_pretty5.default)({
  colorize: true,
  minimumLevel: "info"
}));
var RRWEB_SCRIPT_PATH = (() => {
  const standardPath = import_path2.default.join(getDirname(), "..", "assets", "rrweb", "rrweb.min.js");
  const fallbackPaths = [
    import_path2.default.join(getDirname(), "..", "..", "assets", "rrweb", "rrweb.min.js"),
    // For tests
    import_path2.default.join(process.cwd(), "assets", "rrweb", "rrweb.min.js"),
    // Using cwd
    import_path2.default.join(process.cwd(), "ts-sdk", "assets", "rrweb", "rrweb.min.js")
    // Absolute path
  ];
  try {
    if ((0, import_fs.existsSync)(standardPath)) {
      return standardPath;
    }
    for (const fallbackPath of fallbackPaths) {
      if ((0, import_fs.existsSync)(fallbackPath)) {
        return fallbackPath;
      }
    }
    return standardPath;
  } catch {
    return standardPath;
  }
})();
var PlaywrightInstrumentation = class extends import_instrumentation.InstrumentationBase {
  constructor(client) {
    super(
      "@lmnr/playwright-instrumentation",
      version,
      {
        enabled: true
      }
    );
    this._patchedBrowsers = /* @__PURE__ */ new Set();
    this._patchedContexts = /* @__PURE__ */ new Set();
    this._patchedPages = /* @__PURE__ */ new Set();
    this._parentSpan = void 0;
    this._client = client;
  }
  // It's the caller's responsibility to ensure the span is ended
  setParentSpan(span) {
    this._parentSpan = span;
  }
  init() {
    const module2 = new import_instrumentation.InstrumentationNodeModuleDefinition(
      "playwright",
      // TODO: test if the older versions work
      [">=1.0.0"],
      this.patch.bind(this),
      this.unpatch.bind(this)
    );
    return module2;
  }
  manuallyInstrument(pwModule) {
    const browsers = [pwModule.chromium, pwModule.firefox, pwModule.webkit];
    for (const browserType of browsers) {
      if (browserType) {
        this._wrap(
          browserType,
          "launch",
          this.patchNewBrowser()
        );
        this._wrap(
          browserType,
          "connect",
          this.patchNewBrowser()
        );
        this._wrap(
          browserType,
          "connectOverCDP",
          this.patchNewBrowser()
        );
        this._wrap(
          browserType,
          "launchPersistentContext",
          this.patchBrowserNewContext()
        );
      }
    }
    return pwModule;
  }
  patch(moduleExports, moduleVersion) {
    import_api5.diag.debug(`patching playwright ${moduleVersion}`);
    const browsers = ["chromium", "firefox", "webkit"];
    for (const browserType of browsers) {
      if (moduleExports[browserType]) {
        this._wrap(
          moduleExports[browserType],
          `launch`,
          this.patchNewBrowser()
        );
        this._wrap(
          moduleExports[browserType],
          `connect`,
          this.patchNewBrowser()
        );
        this._wrap(
          moduleExports[browserType],
          "connectOverCDP",
          this.patchNewBrowser()
        );
        this._wrap(
          moduleExports[browserType],
          "launchPersistentContext",
          this.patchBrowserNewContext()
        );
      }
    }
    return moduleExports;
  }
  unpatch(moduleExports, moduleVersion) {
    import_api5.diag.debug(`unpatching playwright ${moduleVersion}`);
    const browsers = ["chromium", "firefox", "webkit"];
    for (const browserType of browsers) {
      if (moduleExports[browserType]) {
        this._wrap(
          moduleExports[browserType],
          `launch`,
          this.patchNewBrowser()
        );
      }
    }
    for (const browser of this._patchedBrowsers) {
      this._unwrap(browser, "newContext");
      this._unwrap(browser, "newPage");
    }
    for (const context4 of this._patchedContexts) {
      this._unwrap(context4, "newPage");
    }
    for (const page of this._patchedPages) {
      this._unwrap(page, "close");
    }
  }
  patchNewBrowser() {
    const plugin = this;
    return (original) => async function method(...args) {
      const browser = await original.call(this, ...args);
      if (!plugin._parentSpan) {
        plugin._parentSpan = Laminar.startSpan({
          name: "playwright"
        });
      }
      for (const context4 of browser.contexts()) {
        plugin._wrap(
          context4,
          "newPage",
          plugin.patchBrowserContextNewPage()
        );
        await Promise.all(context4.pages().map((page) => plugin.patchPage(page)));
        context4.on("page", (page) => plugin.patchPage(page));
        plugin._patchedContexts.add(context4);
      }
      plugin._wrap(
        browser,
        "newContext",
        plugin.patchBrowserNewContext()
      );
      plugin._wrap(
        browser,
        "newPage",
        plugin.patchBrowserNewPage()
      );
      plugin._wrap(
        browser,
        "close",
        plugin.patchBrowserClose()
      );
      plugin._patchedBrowsers.add(browser);
      return browser;
    };
  }
  patchBrowserClose() {
    const plugin = this;
    return (original) => async function method(...args) {
      await original.call(this, ...args);
      if (plugin._parentSpan?.isRecording()) {
        plugin._parentSpan?.end();
      }
    };
  }
  patchBrowserNewContext() {
    const plugin = this;
    return (original) => async function method(...args) {
      const context4 = await original.bind(this).apply(this, args);
      if (!plugin._parentSpan) {
        plugin._parentSpan = Laminar.startSpan({
          name: "playwright"
        });
      }
      plugin._wrap(
        context4,
        "newPage",
        plugin.patchBrowserContextNewPage()
      );
      plugin._wrap(
        context4,
        "close",
        plugin.patchBrowserContextClose()
      );
      context4.on("page", async (page) => plugin.patchPage(page));
      for (const page of context4.pages()) {
        await plugin.patchPage(page);
      }
      plugin._patchedContexts.add(context4);
      return context4;
    };
  }
  patchBrowserContextClose() {
    const plugin = this;
    return (original) => async function method(...args) {
      await original.bind(this).apply(this, args);
      if (plugin._parentSpan?.isRecording()) {
        plugin._parentSpan.end();
      }
    };
  }
  patchBrowserNewPage() {
    return (original) => async function method(...args) {
      const page = await original.bind(this).apply(this, args);
      if (!page) {
        return null;
      }
      return page;
    };
  }
  patchBrowserContextNewPage() {
    const plugin = this;
    return (original) => async function method(...args) {
      const page = await original.bind(this).apply(this, args);
      if (!plugin._parentSpan) {
        plugin._parentSpan = Laminar.startSpan({
          name: "playwright"
        });
      }
      if (!page) {
        return null;
      }
      await plugin.patchPage(page);
      return page;
    };
  }
  async patchPage(page) {
    return await Laminar.withSpan(this._parentSpan, async () => {
      await this._patchPage(page);
    });
  }
  async _patchPage(page) {
    page.addListener("domcontentloaded", (p) => {
      this.injectRrweb(p).catch((error) => {
        logger5.error(`Failed to inject rrweb: ${error instanceof Error ? error.message : String(error)}`);
      });
    });
    await this.injectRrweb(page);
    this._patchedPages.add(page);
    import_api5.trace.getActiveSpan()?.setAttribute(TRACE_HAS_BROWSER_SESSION, true);
    const traceId = import_api5.trace.getActiveSpan()?.spanContext().traceId ?? NIL_UUID;
    const sessionId = newUUID();
    const interval = setInterval(() => {
      if (page.isClosed()) {
        clearInterval(interval);
        return;
      }
      collectAndSendPageEvents(this._client, page, sessionId, traceId).catch((error) => {
        logger5.error(`Event collection stopped: ${error instanceof Error ? error.message : String(error)}`);
      });
    }, 2e3);
    page.addListener("close", async () => {
      clearInterval(interval);
      await collectAndSendPageEvents(this._client, page, sessionId, traceId);
    });
  }
  async injectRrweb(page) {
    await page.waitForLoadState("domcontentloaded");
    const tryRunScript = async (script, maxAttempts = 5) => {
      for (let i = 0; i < maxAttempts; i++) {
        try {
          return await script();
        } catch (error) {
          logger5.error("Operation " + script.name + ` failed: ${error instanceof Error ? error.message : String(error)}`);
        }
        await new Promise((resolve) => setTimeout(resolve, 500));
      }
    };
    const isRrwebPresent = await page.evaluate(
      () => typeof window.lmnrRrweb !== "undefined"
    );
    if (!isRrwebPresent) {
      const script = await (0, import_promises.readFile)(RRWEB_SCRIPT_PATH, "utf8");
      await tryRunScript(async function injectRrweb() {
        await page.evaluate(script);
        const res = await page.waitForFunction(
          () => "lmnrRrweb" in window || window.lmnrRrweb,
          {
            timeout: 5e3
          }
        );
        if (!res) {
          throw new Error("Failed to inject rrweb");
        }
      });
    }
    await tryRunScript(async function setupRrwebCollection() {
      await page.evaluate(() => {
        const BATCH_SIZE = 1e3;
        const HEARTBEAT_INTERVAL = 1e3;
        window.lmnrRrwebEventsBatch = /* @__PURE__ */ new Set();
        const compressEventData = async (data) => {
          const jsonString = JSON.stringify(data);
          const blob = new Blob([jsonString], { type: "application/json" });
          const compressedStream = blob.stream().pipeThrough(new CompressionStream("gzip"));
          const compressedResponse = new Response(compressedStream);
          const compressedData = await compressedResponse.arrayBuffer();
          return Array.from(new Uint8Array(compressedData));
        };
        window.lmnrGetAndClearEvents = () => {
          const events = window.lmnrRrwebEventsBatch;
          window.lmnrRrwebEventsBatch = /* @__PURE__ */ new Set();
          return Array.from(events);
        };
        setInterval(() => {
          compressEventData({ source: "heartbeat" }).then((data) => {
            const heartbeatEvent = {
              type: 6,
              // Custom event type
              data,
              timestamp: Date.now()
            };
            window.lmnrRrwebEventsBatch?.add(heartbeatEvent);
            if (window.lmnrRrwebEventsBatch?.length > BATCH_SIZE) {
              window.lmnrRrwebEventsBatch = new Set(
                Array.from(window.lmnrRrwebEventsBatch).slice(-BATCH_SIZE)
              );
            }
          }).catch((error) => {
            logger5.error(`Heartbeat failed: ${error instanceof Error ? error.message : String(error)}`);
          });
        }, HEARTBEAT_INTERVAL);
        window.lmnrRrweb.record({
          async emit(event) {
            if (document.visibilityState === "hidden" || document.hidden) {
              return;
            }
            const compressedEvent = {
              ...event,
              data: await compressEventData(event.data)
            };
            window.lmnrRrwebEventsBatch.add(compressedEvent);
          }
        });
      });
    });
  }
};

// src/browser/stagehand.ts
var import_api6 = require("@opentelemetry/api");
var import_instrumentation2 = require("@opentelemetry/instrumentation");
var StagehandInstrumentation = class extends import_instrumentation2.InstrumentationBase {
  constructor(playwrightInstrumentation2) {
    super(
      "@lmnr/browserbase-stagehand-instrumentation",
      version,
      {
        enabled: true
      }
    );
    this.playwrightInstrumentation = playwrightInstrumentation2;
  }
  init() {
    const module2 = new import_instrumentation2.InstrumentationNodeModuleDefinition(
      "@browserbasehq/stagehand",
      [">=1.0.0"],
      this.patch.bind(this),
      this.unpatch.bind(this)
    );
    return module2;
  }
  patch(moduleExports, moduleVersion) {
    import_api6.diag.debug(`patching stagehand ${moduleVersion}`);
    const descriptor = Object.getOwnPropertyDescriptor(moduleExports, "Stagehand");
    if (descriptor && !descriptor.configurable) {
      const originalStagehand = moduleExports.Stagehand;
      const patchedConstructor = this.patchStagehandConstructor()(originalStagehand);
      return new Proxy(moduleExports, {
        get: (target, prop) => {
          if (prop === "Stagehand") {
            return patchedConstructor;
          }
          return target[prop];
        }
      });
    } else {
      this._wrap(
        moduleExports,
        "Stagehand",
        this.patchStagehandConstructor()
      );
      return moduleExports;
    }
  }
  manuallyInstrument(Stagehand) {
    import_api6.diag.debug("manually instrumenting stagehand");
    if (Stagehand && Stagehand.prototype) {
      this._wrap(
        Stagehand.prototype,
        "init",
        this.patchStagehandInit()
      );
      this._wrap(
        Stagehand.prototype,
        "close",
        this.patchStagehandClose()
      );
    }
  }
  unpatch(moduleExports, moduleVersion) {
    import_api6.diag.debug(`unpatching stagehand ${moduleVersion}`);
    this._unwrap(moduleExports, "Stagehand");
    if (moduleExports.Stagehand) {
      this._unwrap(moduleExports.Stagehand.prototype, "init");
      this._unwrap(moduleExports.Stagehand.prototype, "close");
      if (moduleExports.Stagehand.prototype?.page) {
        this._unwrap(moduleExports.Stagehand.prototype.page, "act");
        this._unwrap(moduleExports.Stagehand.prototype.page, "extract");
        this._unwrap(moduleExports.Stagehand.prototype.page, "observe");
        const observeHandler = moduleExports.Stagehand.prototype.page.observeHandler;
        if (observeHandler) {
          this._unwrap(observeHandler, "observe");
        }
        const extractHandler = moduleExports.Stagehand.prototype.page.extractHandler;
        if (extractHandler) {
          this._unwrap(extractHandler, "textExtract");
          this._unwrap(extractHandler, "domExtract");
        }
        const actHandler = moduleExports.Stagehand.prototype.page.actHandler;
        if (actHandler) {
          this._unwrap(actHandler, "act");
        }
      }
    }
    return moduleExports;
  }
  patchStagehandConstructor() {
    const instrumentation = this;
    return (Original) => {
      const Stagehand = function(...args) {
        if (!(this instanceof Stagehand)) {
          return new Stagehand(...args);
        }
        const instance = new Original(args.length > 0 ? args[0] : void 0);
        Object.assign(this, instance);
        instrumentation._wrap(
          this,
          "init",
          instrumentation.patchStagehandInit()
        );
        instrumentation._wrap(
          this,
          "close",
          instrumentation.patchStagehandClose()
        );
        return this;
      };
      Object.setPrototypeOf(Stagehand, Original);
      Stagehand.prototype = Object.create(Original.prototype);
      Stagehand.prototype.constructor = Stagehand;
      return Stagehand;
    };
  }
  patchStagehandInit() {
    const instrumentation = this;
    return (original) => async function method(...args) {
      instrumentation._parentSpan = Laminar.startSpan({
        name: "Stagehand"
      });
      instrumentation.playwrightInstrumentation.setParentSpan(instrumentation._parentSpan);
      const result = await original.bind(this).apply(this, args);
      await instrumentation.playwrightInstrumentation.patchPage(this.page);
      instrumentation.patchStagehandPage(this.stagehandPage);
      return result;
    };
  }
  patchStagehandClose() {
    const instrumentation = this;
    return (original) => async function method(...args) {
      if (instrumentation._parentSpan && instrumentation._parentSpan.isRecording()) {
        instrumentation._parentSpan.end();
      }
      await original.bind(this).apply(this, args);
    };
  }
  patchStagehandPage(page) {
    const actHandler = page.actHandler;
    if (actHandler) {
      this._wrap(
        actHandler,
        "act",
        this.patchStagehandActHandler()
      );
    }
    const observeHandler = page.observeHandler;
    if (observeHandler) {
      this._wrap(
        observeHandler,
        "observe",
        this.patchStagehandObserveHandler()
      );
    }
    const extractHandler = page.extractHandler;
    if (extractHandler) {
      this._wrap(
        extractHandler,
        "textExtract",
        this.patchStagehandExtractHandlerTextExtract()
      );
      this._wrap(
        extractHandler,
        "domExtract",
        this.patchStagehandExtractHandlerDomExtract()
      );
    }
    this._wrap(
      page,
      "act",
      this.patchStagehandGlobalMethod("act")
    );
    this._wrap(
      page,
      "extract",
      this.patchStagehandGlobalMethod("extract")
    );
    this._wrap(
      page,
      "observe",
      this.patchStagehandGlobalMethod("observe")
    );
  }
  patchStagehandGlobalMethod(methodName) {
    const instrumentation = this;
    return (original) => async function method(...args) {
      return await Laminar.withSpan(
        instrumentation._parentSpan,
        async () => await observe(
          { name: `stagehand.${methodName}`, input: args },
          async (thisArg, ...rest) => await original.apply(thisArg, ...rest),
          this,
          args
        )
      );
    };
  }
  patchStagehandActHandler() {
    return (original) => async function act(...args) {
      return await observe(
        {
          name: "stagehand.actHandler.act",
          input: {
            action: args[0].action ?? null,
            llmClient: cleanStagehandLLMClient(args[0].llmClient ?? {}),
            chunksSeen: args[0].chunksSeen ?? null,
            steps: args[0].steps ?? null,
            requestId: args[0].requestId ?? null,
            schema: args[0].schema ?? null,
            retries: args[0].retries ?? null,
            variables: args[0].variables ?? null,
            previousSelectors: args[0].previousSelectors ?? null,
            skipActionCacheForThisStep: args[0].skipActionCacheForThisStep ?? null,
            domSettleTimeoutMs: args[0].domSettleTimeoutMs ?? null
          }
        },
        async () => await original.bind(this).apply(this, args)
      );
    };
  }
  patchStagehandExtractHandlerTextExtract() {
    return (original) => async function textExtract(...args) {
      return await observe(
        {
          name: "stagehand.extractHandler.textExtract",
          input: {
            instruction: args[0].instruction ?? null,
            llmClient: cleanStagehandLLMClient(args[0].llmClient ?? {}),
            requestId: args[0].requestId ?? null,
            schema: args[0].schema ?? null,
            content: args[0].content ?? null,
            domSettleTimeoutMs: args[0].domSettleTimeoutMs ?? null
          }
        },
        async () => await original.bind(this).apply(this, args)
      );
    };
  }
  patchStagehandExtractHandlerDomExtract() {
    return (original) => async function domExtract(...args) {
      return await observe(
        {
          name: "stagehand.extractHandler.domExtract",
          input: {
            instruction: args[0].instruction ?? null,
            llmClient: cleanStagehandLLMClient(args[0].llmClient ?? {}),
            requestId: args[0].requestId ?? null,
            schema: args[0].schema ?? null,
            content: args[0].content ?? null,
            chunksSeen: args[0].chunksSeen ?? null,
            domSettleTimeoutMs: args[0].domSettleTimeoutMs ?? null
          }
        },
        async () => await original.apply(this, args)
      );
    };
  }
  patchStagehandObserveHandler() {
    return (original) => async function observe2(...args) {
      return await observe(
        {
          name: "stagehand.observeHandler.observe",
          input: {
            instruction: args[0].instruction ?? null,
            llmClient: cleanStagehandLLMClient(args[0].llmClient ?? {}),
            requestId: args[0].requestId ?? null,
            returnAction: args[0].returnAction ?? null,
            onlyVisible: args[0].onlyVisible ?? null,
            drawOverlay: args[0].drawOverlay ?? null
          }
        },
        async () => await original.bind(this).apply(this, args)
      );
    };
  }
};

// src/browser/puppeteer.ts
var import_api7 = require("@opentelemetry/api");
var import_instrumentation3 = require("@opentelemetry/instrumentation");
var import_fs2 = require("fs");
var import_promises2 = require("fs/promises");
var import_path3 = __toESM(require("path"));
var import_pino6 = __toESM(require("pino"));
var import_pino_pretty6 = __toESM(require("pino-pretty"));
var logger6 = (0, import_pino6.default)((0, import_pino_pretty6.default)({
  colorize: true,
  minimumLevel: "info"
}));
var RRWEB_SCRIPT_PATH2 = (() => {
  const standardPath = import_path3.default.join(getDirname(), "..", "assets", "rrweb", "rrweb.min.js");
  const fallbackPaths = [
    import_path3.default.join(getDirname(), "..", "..", "assets", "rrweb", "rrweb.min.js"),
    // For tests
    import_path3.default.join(process.cwd(), "assets", "rrweb", "rrweb.min.js"),
    // Using cwd
    import_path3.default.join(process.cwd(), "ts-sdk", "assets", "rrweb", "rrweb.min.js")
    // Absolute path
  ];
  try {
    if ((0, import_fs2.existsSync)(standardPath)) {
      return standardPath;
    }
    for (const fallbackPath of fallbackPaths) {
      if ((0, import_fs2.existsSync)(fallbackPath)) {
        return fallbackPath;
      }
    }
    return standardPath;
  } catch {
    return standardPath;
  }
})();
var PuppeteerInstrumentation = class extends import_instrumentation3.InstrumentationBase {
  constructor(client) {
    super(
      "@lmnr/puppeteer-instrumentation",
      version,
      {
        enabled: true
      }
    );
    this._patchedBrowsers = /* @__PURE__ */ new Set();
    this._patchedContexts = /* @__PURE__ */ new Set();
    this._patchedPages = /* @__PURE__ */ new Set();
    this._parentSpan = void 0;
    this._client = client;
  }
  // It's the caller's responsibility to ensure the span is ended
  setParentSpan(span) {
    this._parentSpan = span;
  }
  init() {
    const puppeteerInstrumentation2 = new import_instrumentation3.InstrumentationNodeModuleDefinition(
      "puppeteer",
      // About two years before first writing this instrumentation
      // and apparently no big breaking changes afterwards
      // https://github.com/puppeteer/puppeteer/releases
      [">=19.0.0"],
      this.patch.bind(this),
      this.unpatch.bind(this)
    );
    const puppeteerCoreInstrumentation = new import_instrumentation3.InstrumentationNodeModuleDefinition(
      "puppeteer-core",
      [">=19.0.0"],
      this.patch.bind(this),
      this.unpatch.bind(this)
    );
    return [puppeteerInstrumentation2, puppeteerCoreInstrumentation];
  }
  manuallyInstrument(puppeteerModule) {
    this._wrap(
      puppeteerModule,
      "launch",
      this.patchNewBrowser()
    );
    this._wrap(
      puppeteerModule,
      "connect",
      this.patchNewBrowser()
    );
    return puppeteerModule;
  }
  patch(moduleExports, moduleVersion) {
    import_api7.diag.debug(`patching puppeteer ${moduleVersion}`);
    this._wrap(
      moduleExports,
      `launch`,
      this.patchNewBrowser()
    );
    this._wrap(
      moduleExports.PuppeteerNode.prototype,
      `launch`,
      this.patchNewBrowser()
    );
    this._wrap(
      moduleExports,
      `connect`,
      this.patchNewBrowser()
    );
    this._wrap(
      moduleExports.Puppeteer.prototype,
      `connect`,
      this.patchNewBrowser()
    );
    this._wrap(
      moduleExports.default,
      `launch`,
      this.patchNewBrowser()
    );
    this._wrap(
      moduleExports.default,
      `connect`,
      this.patchNewBrowser()
    );
    return moduleExports;
  }
  unpatch(moduleExports, moduleVersion) {
    import_api7.diag.debug(`unpatching puppeteer ${moduleVersion}`);
    this._unwrap(
      moduleExports,
      `launch`
    );
    this._unwrap(
      moduleExports,
      `connect`
    );
    this._unwrap(
      moduleExports.PuppeteerNode.prototype,
      `launch`
    );
    this._unwrap(
      moduleExports.Puppeteer.prototype,
      `connect`
    );
    this._unwrap(
      moduleExports.default,
      `connect`
    );
    this._unwrap(
      moduleExports.default,
      `launch`
    );
  }
  patchNewBrowser() {
    const plugin = this;
    return (original) => async function method(...args) {
      const browser = await original.call(this, ...args);
      if (!plugin._parentSpan) {
        plugin._parentSpan = Laminar.startSpan({
          name: "puppeteer"
        });
      }
      for (const context4 of browser.browserContexts()) {
        plugin._wrap(
          context4,
          "newPage",
          plugin.patchBrowserContextNewPage()
        );
        await Promise.all((await context4.pages()).map((page) => plugin.patchPage(page)));
        context4.on("page", (page) => {
          plugin.patchPage(page).catch((error) => {
            logger6.error(`Failed to patch page: ${error instanceof Error ? error.message : String(error)}`);
          });
        });
        plugin._patchedContexts.add(context4);
      }
      plugin._wrap(
        browser,
        "createBrowserContext",
        plugin.patchBrowserNewContext()
      );
      plugin._wrap(
        browser,
        "newPage",
        plugin.patchBrowserNewPage()
      );
      plugin._wrap(
        browser,
        "close",
        plugin.patchBrowserClose()
      );
      plugin._patchedBrowsers.add(browser);
      return browser;
    };
  }
  patchBrowserClose() {
    const plugin = this;
    return (original) => async function method(...args) {
      await original.call(this, ...args);
      if (plugin._parentSpan?.isRecording()) {
        plugin._parentSpan?.end();
      }
    };
  }
  patchBrowserNewContext() {
    const plugin = this;
    return (original) => async function method(...args) {
      const context4 = await original.bind(this).apply(this, args);
      if (!plugin._parentSpan) {
        plugin._parentSpan = Laminar.startSpan({
          name: "puppeteer"
        });
      }
      plugin._wrap(
        context4,
        "newPage",
        plugin.patchBrowserContextNewPage()
      );
      plugin._wrap(
        context4,
        "close",
        plugin.patchBrowserContextClose()
      );
      context4.on("page", (page) => {
        plugin.patchPage(page).catch((error) => {
          logger6.error(`Failed to patch page: ${error instanceof Error ? error.message : String(error)}`);
        });
      });
      for (const page of await context4.pages()) {
        await plugin.patchPage(page);
      }
      plugin._patchedContexts.add(context4);
      return context4;
    };
  }
  patchBrowserNewContextSync() {
    const plugin = this;
    return (original) => function method(...args) {
      const context4 = original.bind(this).apply(this, args);
      if (!plugin._parentSpan) {
        plugin._parentSpan = Laminar.startSpan({
          name: "puppeteer"
        });
      }
      plugin._wrap(
        context4,
        "newPage",
        plugin.patchBrowserContextNewPage()
      );
      plugin._wrap(
        context4,
        "close",
        plugin.patchBrowserContextClose()
      );
      context4.on("page", (page) => {
        plugin.patchPage(page).catch((error) => {
          logger6.error(`Failed to patch page: ${error instanceof Error ? error.message : String(error)}`);
        });
      });
      context4.pages().then((pages) => {
        for (const page of pages) {
          plugin.patchPage(page).catch((error) => {
            logger6.error(`Failed to patch page: ${error instanceof Error ? error.message : String(error)}`);
          });
        }
      }).catch((error) => {
        logger6.error(`Failed to patch pages: ${error instanceof Error ? error.message : String(error)}`);
      });
      plugin._patchedContexts.add(context4);
      return context4;
    };
  }
  patchBrowserContextClose() {
    const plugin = this;
    return (original) => async function method(...args) {
      await original.bind(this).apply(this, args);
      if (plugin._parentSpan?.isRecording()) {
        plugin._parentSpan.end();
      }
    };
  }
  patchBrowserNewPage() {
    return (original) => async function method(...args) {
      const page = await original.bind(this).apply(this, args);
      if (!page) {
        return null;
      }
      return page;
    };
  }
  patchBrowserContextNewPage() {
    const plugin = this;
    return (original) => async function method(...args) {
      const page = await original.bind(this).apply(this, args);
      if (!plugin._parentSpan) {
        plugin._parentSpan = Laminar.startSpan({
          name: "puppeteer"
        });
      }
      if (!page) {
        return null;
      }
      await plugin.patchPage(page);
      return page;
    };
  }
  async patchPage(page) {
    return await Laminar.withSpan(this._parentSpan, async () => {
      await this._patchPage(page);
    });
  }
  async _patchPage(page) {
    page.on("domcontentloaded", () => {
      this.injectRrweb(page).catch((error) => {
        logger6.error(`Failed to inject rrweb: ${error instanceof Error ? error.message : String(error)}`);
      });
    });
    await this.injectRrweb(page);
    this._patchedPages.add(page);
    import_api7.trace.getActiveSpan()?.setAttribute(TRACE_HAS_BROWSER_SESSION, true);
    const traceId = import_api7.trace.getActiveSpan()?.spanContext().traceId ?? NIL_UUID;
    const sessionId = newUUID();
    const interval = setInterval(() => {
      if (page.isClosed()) {
        clearInterval(interval);
        return;
      }
      collectAndSendPageEvents(this._client, page, sessionId, traceId).catch((error) => {
        logger6.error(`Event collection stopped: ${error instanceof Error ? error.message : String(error)}`);
      });
    }, 2e3);
    page.on("close", () => {
      clearInterval(interval);
      collectAndSendPageEvents(this._client, page, sessionId, traceId).catch((error) => {
        logger6.error(`Event collection stopped: ${error instanceof Error ? error.message : String(error)}`);
      });
    });
  }
  async injectRrweb(page) {
    const tryRunScript = async (script, maxAttempts = 5) => {
      for (let i = 0; i < maxAttempts; i++) {
        try {
          return await script();
        } catch (error) {
          const errorMessage = error instanceof Error ? error.message : String(error);
          if (errorMessage.includes("Execution context was destroyed")) {
            logger6.info(errorMessage);
          } else {
            logger6.error("Operation " + script.name + " failed: " + errorMessage);
          }
        }
        await new Promise((resolve) => setTimeout(resolve, 500));
      }
    };
    const isRrwebPresent = await page.evaluate(
      () => typeof window.lmnrRrweb !== "undefined"
    );
    if (!isRrwebPresent) {
      const script = await (0, import_promises2.readFile)(RRWEB_SCRIPT_PATH2, "utf8");
      await tryRunScript(async function injectRrweb() {
        await page.evaluate(script);
        const res = await page.waitForFunction(
          () => "lmnrRrweb" in window || window.lmnrRrweb,
          {
            timeout: 5e3
          }
        );
        if (!res) {
          throw new Error("Failed to inject rrweb");
        }
      });
    }
    await tryRunScript(async function setupRrwebCollection() {
      await page.evaluate(() => {
        const BATCH_SIZE = 1e3;
        const HEARTBEAT_INTERVAL = 1e3;
        window.lmnrRrwebEventsBatch = /* @__PURE__ */ new Set();
        const compressEventData = async (data) => {
          const jsonString = JSON.stringify(data);
          const blob = new Blob([jsonString], { type: "application/json" });
          const compressedStream = blob.stream().pipeThrough(new CompressionStream("gzip"));
          const compressedResponse = new Response(compressedStream);
          const compressedData = await compressedResponse.arrayBuffer();
          return Array.from(new Uint8Array(compressedData));
        };
        window.lmnrGetAndClearEvents = () => {
          const events = window.lmnrRrwebEventsBatch;
          window.lmnrRrwebEventsBatch = /* @__PURE__ */ new Set();
          return Array.from(events);
        };
        setInterval(() => {
          compressEventData({ source: "heartbeat" }).then((data) => {
            const heartbeatEvent = {
              type: 6,
              // Custom event type
              data,
              timestamp: Date.now()
            };
            window.lmnrRrwebEventsBatch?.add(heartbeatEvent);
            if (window.lmnrRrwebEventsBatch?.length > BATCH_SIZE) {
              window.lmnrRrwebEventsBatch = new Set(
                Array.from(window.lmnrRrwebEventsBatch).slice(-BATCH_SIZE)
              );
            }
          }).catch((error) => {
            logger6.error(`Heartbeat failed: ${error instanceof Error ? error.message : String(error)}`);
          });
        }, HEARTBEAT_INTERVAL);
        window.lmnrRrweb.record({
          async emit(event) {
            if (document.visibilityState === "hidden" || document.hidden) {
              return;
            }
            const compressedEvent = {
              ...event,
              data: await compressEventData(event.data)
            };
            window.lmnrRrwebEventsBatch.add(compressedEvent);
          }
        });
      });
    });
  }
};

// src/sdk/tracing/index.ts
var logger7 = (0, import_pino7.default)((0, import_pino_pretty7.default)({
  colorize: true,
  minimumLevel: "info"
}));
var _spanProcessor;
var _spanIdToPath = /* @__PURE__ */ new Map();
var _spanIdLists = /* @__PURE__ */ new Map();
var openAIInstrumentation;
var anthropicInstrumentation;
var azureOpenAIInstrumentation;
var cohereInstrumentation;
var vertexaiInstrumentation;
var aiplatformInstrumentation;
var bedrockInstrumentation;
var langchainInstrumentation;
var llamaIndexInstrumentation;
var pineconeInstrumentation;
var chromadbInstrumentation;
var qdrantInstrumentation;
var playwrightInstrumentation;
var stagehandInstrumentation;
var puppeteerInstrumentation;
var togetherInstrumentation;
var NUM_TRACKED_NEXTJS_SPANS = 1e4;
var instrumentations = [];
var initInstrumentations = (client) => {
  const enrichTokens = false;
  openAIInstrumentation = new import_instrumentation_openai.OpenAIInstrumentation({
    enrichTokens
  });
  instrumentations.push(openAIInstrumentation);
  anthropicInstrumentation = new import_instrumentation_anthropic.AnthropicInstrumentation();
  instrumentations.push(anthropicInstrumentation);
  azureOpenAIInstrumentation = new import_instrumentation_azure.AzureOpenAIInstrumentation();
  instrumentations.push(azureOpenAIInstrumentation);
  cohereInstrumentation = new import_instrumentation_cohere.CohereInstrumentation();
  instrumentations.push(cohereInstrumentation);
  vertexaiInstrumentation = new import_instrumentation_vertexai.VertexAIInstrumentation();
  instrumentations.push(vertexaiInstrumentation);
  aiplatformInstrumentation = new import_instrumentation_vertexai.AIPlatformInstrumentation();
  instrumentations.push(aiplatformInstrumentation);
  bedrockInstrumentation = new import_instrumentation_bedrock.BedrockInstrumentation();
  instrumentations.push(bedrockInstrumentation);
  pineconeInstrumentation = new import_instrumentation_pinecone.PineconeInstrumentation();
  instrumentations.push(pineconeInstrumentation);
  langchainInstrumentation = new import_instrumentation_langchain.LangChainInstrumentation();
  instrumentations.push(langchainInstrumentation);
  llamaIndexInstrumentation = new import_instrumentation_llamaindex.LlamaIndexInstrumentation();
  instrumentations.push(llamaIndexInstrumentation);
  chromadbInstrumentation = new import_instrumentation_chromadb.ChromaDBInstrumentation();
  instrumentations.push(chromadbInstrumentation);
  qdrantInstrumentation = new import_instrumentation_qdrant.QdrantInstrumentation();
  instrumentations.push(qdrantInstrumentation);
  playwrightInstrumentation = new PlaywrightInstrumentation(client);
  instrumentations.push(playwrightInstrumentation);
  stagehandInstrumentation = new StagehandInstrumentation(playwrightInstrumentation);
  instrumentations.push(stagehandInstrumentation);
  puppeteerInstrumentation = new PuppeteerInstrumentation(client);
  instrumentations.push(puppeteerInstrumentation);
  togetherInstrumentation = new import_instrumentation_together.TogetherInstrumentation();
  instrumentations.push(togetherInstrumentation);
};
var manuallyInitInstrumentations = (client, instrumentModules) => {
  const enrichTokens = false;
  instrumentations.length = 0;
  if (instrumentModules?.openAI) {
    openAIInstrumentation = new import_instrumentation_openai.OpenAIInstrumentation({
      enrichTokens
    });
    instrumentations.push(openAIInstrumentation);
    openAIInstrumentation.manuallyInstrument(instrumentModules.openAI);
  }
  if (instrumentModules?.anthropic) {
    anthropicInstrumentation = new import_instrumentation_anthropic.AnthropicInstrumentation();
    instrumentations.push(anthropicInstrumentation);
    anthropicInstrumentation.manuallyInstrument(instrumentModules.anthropic);
  }
  if (instrumentModules?.azureOpenAI) {
    const instrumentation = new import_instrumentation_azure.AzureOpenAIInstrumentation();
    instrumentations.push(instrumentation);
    azureOpenAIInstrumentation = instrumentation;
    instrumentation.manuallyInstrument(instrumentModules.azureOpenAI);
  }
  if (instrumentModules?.cohere) {
    cohereInstrumentation = new import_instrumentation_cohere.CohereInstrumentation();
    instrumentations.push(cohereInstrumentation);
    cohereInstrumentation.manuallyInstrument(instrumentModules.cohere);
  }
  if (instrumentModules?.google_vertexai) {
    vertexaiInstrumentation = new import_instrumentation_vertexai.VertexAIInstrumentation();
    instrumentations.push(vertexaiInstrumentation);
    vertexaiInstrumentation.manuallyInstrument(
      instrumentModules.google_vertexai
    );
  }
  if (instrumentModules?.google_aiplatform) {
    aiplatformInstrumentation = new import_instrumentation_vertexai.AIPlatformInstrumentation();
    instrumentations.push(aiplatformInstrumentation);
    aiplatformInstrumentation.manuallyInstrument(
      instrumentModules.google_aiplatform
    );
  }
  if (instrumentModules?.bedrock) {
    bedrockInstrumentation = new import_instrumentation_bedrock.BedrockInstrumentation();
    instrumentations.push(bedrockInstrumentation);
    bedrockInstrumentation.manuallyInstrument(instrumentModules.bedrock);
  }
  if (instrumentModules?.pinecone) {
    const instrumentation = new import_instrumentation_pinecone.PineconeInstrumentation();
    instrumentations.push(instrumentation);
    instrumentation.manuallyInstrument(instrumentModules.pinecone);
  }
  if (instrumentModules?.langchain) {
    langchainInstrumentation = new import_instrumentation_langchain.LangChainInstrumentation();
    instrumentations.push(langchainInstrumentation);
    langchainInstrumentation.manuallyInstrument(instrumentModules.langchain);
  }
  if (instrumentModules?.llamaIndex) {
    llamaIndexInstrumentation = new import_instrumentation_llamaindex.LlamaIndexInstrumentation();
    instrumentations.push(llamaIndexInstrumentation);
    llamaIndexInstrumentation.manuallyInstrument(instrumentModules.llamaIndex);
  }
  if (instrumentModules?.chromadb) {
    chromadbInstrumentation = new import_instrumentation_chromadb.ChromaDBInstrumentation();
    instrumentations.push(chromadbInstrumentation);
    chromadbInstrumentation.manuallyInstrument(instrumentModules.chromadb);
  }
  if (instrumentModules?.qdrant) {
    qdrantInstrumentation = new import_instrumentation_qdrant.QdrantInstrumentation();
    instrumentations.push(qdrantInstrumentation);
    qdrantInstrumentation.manuallyInstrument(instrumentModules.qdrant);
  }
  if (instrumentModules?.together) {
    togetherInstrumentation = new import_instrumentation_together.TogetherInstrumentation();
    instrumentations.push(togetherInstrumentation);
    togetherInstrumentation.manuallyInstrument(instrumentModules.together);
  }
  if (instrumentModules?.playwright) {
    playwrightInstrumentation = new PlaywrightInstrumentation(client);
    instrumentations.push(playwrightInstrumentation);
    playwrightInstrumentation.manuallyInstrument(instrumentModules.playwright);
  }
  if (instrumentModules?.puppeteer) {
    puppeteerInstrumentation = new PuppeteerInstrumentation(client);
    instrumentations.push(puppeteerInstrumentation);
    puppeteerInstrumentation.manuallyInstrument(instrumentModules.puppeteer);
  }
  if (instrumentModules?.stagehand) {
    if (!playwrightInstrumentation) {
      playwrightInstrumentation = new PlaywrightInstrumentation(client);
      instrumentations.push(playwrightInstrumentation);
    }
    stagehandInstrumentation = new StagehandInstrumentation(playwrightInstrumentation);
    instrumentations.push(stagehandInstrumentation);
    stagehandInstrumentation.manuallyInstrument(instrumentModules.stagehand);
  }
};
var NEXT_JS_SPAN_ATTRIBUTES = [
  "next.span_name",
  "next.span_type",
  "next.clientComponentLoadCount",
  "next.page",
  "next.rsc",
  "next.route",
  "next.segment"
];
var isNextJsSpan = (span) => NEXT_JS_SPAN_ATTRIBUTES.some((attr) => span.attributes[attr] !== void 0);
var startTracing = (options) => {
  const client = new LaminarClient({
    baseUrl: options.baseUrl ?? "https://api.lmnr.ai",
    projectApiKey: options.apiKey ?? process.env.LMNR_PROJECT_API_KEY
  });
  if (options.instrumentModules !== void 0) {
    manuallyInitInstrumentations(client, options.instrumentModules);
  } else {
    initInstrumentations(client);
  }
  if (!shouldSendTraces()) {
    openAIInstrumentation?.setConfig({
      traceContent: false
    });
    azureOpenAIInstrumentation?.setConfig({
      traceContent: false
    });
    llamaIndexInstrumentation?.setConfig({
      traceContent: false
    });
    vertexaiInstrumentation?.setConfig({
      traceContent: false
    });
    aiplatformInstrumentation?.setConfig({
      traceContent: false
    });
    bedrockInstrumentation?.setConfig({
      traceContent: false
    });
    cohereInstrumentation?.setConfig({
      traceContent: false
    });
    chromadbInstrumentation?.setConfig({
      traceContent: false
    });
  }
  const headers = process.env.TRACELOOP_HEADERS ? import_core.baggageUtils.parseKeyPairsIntoRecord(process.env.TRACELOOP_HEADERS) : { Authorization: `Bearer ${options.apiKey}` };
  const traceExporter = options.exporter ?? new import_exporter_trace_otlp_proto.OTLPTraceExporter({
    url: `${options.baseUrl}/v1/traces`,
    headers
  });
  _spanProcessor = options.processor ?? (options.disableBatch ? new import_sdk_trace_node.SimpleSpanProcessor(traceExporter) : new import_sdk_trace_node.BatchSpanProcessor(traceExporter, {
    maxExportBatchSize: options.maxExportBatchSize
  }));
  const nextJsSpanIds = /* @__PURE__ */ new Set();
  const addNextJsSpanId = (spanId) => {
    if (nextJsSpanIds.size >= NUM_TRACKED_NEXTJS_SPANS) {
      const toRemove = Array.from(nextJsSpanIds).slice(0, NUM_TRACKED_NEXTJS_SPANS / 2);
      toRemove.forEach((id) => nextJsSpanIds.delete(id));
    }
    nextJsSpanIds.add(spanId);
  };
  const originalOnEnd = _spanProcessor.onEnd.bind(_spanProcessor);
  const originalOnStart = _spanProcessor.onStart.bind(_spanProcessor);
  _spanProcessor.onStart = (span, parentContext) => {
    const contextSpanPath = getSpanPath(parentContext ?? import_api8.context.active());
    const parentSpanPath = contextSpanPath ?? (span.parentSpanId !== void 0 ? _spanIdToPath.get(span.parentSpanId) : void 0);
    const spanId = span.spanContext().spanId;
    const parentSpanIdsPath = span.parentSpanId ? !options.preserveNextJsSpans && nextJsSpanIds.has(span.parentSpanId) ? [] : _spanIdLists.get(span.parentSpanId) ?? [] : [];
    const spanPath = !options.preserveNextJsSpans && span.parentSpanId && nextJsSpanIds.has(span.parentSpanId) ? [span.name] : parentSpanPath ? [...parentSpanPath, span.name] : [span.name];
    const spanIdUuid = otelSpanIdToUUID(spanId);
    const spanIdsPath = parentSpanIdsPath ? [...parentSpanIdsPath, spanIdUuid] : [spanIdUuid];
    span.setAttribute(SPAN_IDS_PATH, spanIdsPath);
    _spanIdLists.set(spanId, spanIdsPath);
    span.setAttribute(SPAN_PATH, spanPath);
    import_api8.context.active().setValue(SPAN_PATH_KEY, spanPath);
    _spanIdToPath.set(spanId, spanPath);
    span.setAttribute(SPAN_INSTRUMENTATION_SOURCE, "javascript");
    span.setAttribute(SPAN_SDK_VERSION, version);
    if (getLangVersion()) {
      span.setAttribute(SPAN_LANGUAGE_VERSION, getLangVersion());
    }
    const associationProperties = import_api8.context.active().getValue(ASSOCIATION_PROPERTIES_KEY);
    if (associationProperties) {
      for (const [key, value] of Object.entries(associationProperties)) {
        if (Object.keys(ASSOCIATION_PROPERTIES_OVERRIDES).includes(key)) {
          span.setAttribute(ASSOCIATION_PROPERTIES_OVERRIDES[key], value);
        } else if (key === "tracing_level") {
          span.setAttribute("lmnr.internal.tracing_level", value);
        } else {
          span.setAttribute(`${ASSOCIATION_PROPERTIES}.${key}`, value);
        }
      }
    }
    if (span.parentSpanId && nextJsSpanIds.has(span.parentSpanId) && !options.preserveNextJsSpans) {
      span.setAttribute(OVERRIDE_PARENT_SPAN, true);
      span.setAttribute(EXTRACTED_FROM_NEXT_JS, true);
    }
    originalOnStart(span, parentContext);
    if (isNextJsSpan(span) && !options.preserveNextJsSpans) {
      addNextJsSpanId(span.spanContext().spanId);
    }
  };
  _spanProcessor.onEnd = (span) => {
    if ((isNextJsSpan(span) || nextJsSpanIds.has(span.spanContext().spanId)) && !options.preserveNextJsSpans) {
      addNextJsSpanId(span.spanContext().spanId);
    } else {
      originalOnEnd(span);
    }
  };
  if (options.useExternalTracerProvider) {
    const globalProvider = import_api8.trace.getTracerProvider();
    let provider;
    if (globalProvider instanceof import_api8.ProxyTracerProvider) {
      provider = globalProvider.getDelegate();
    } else {
      provider = globalProvider;
    }
    if (provider.addSpanProcessor) {
      provider.addSpanProcessor(_spanProcessor);
    } else {
      throw new Error("The active tracer provider does not support adding a span processor");
    }
  } else {
    const provider = new import_sdk_trace_node.NodeTracerProvider({
      spanProcessors: [_spanProcessor]
    });
    try {
      provider.register();
    } catch {
      logger7.error("Error registering tracer provider");
    }
    (0, import_instrumentation4.registerInstrumentations)({
      instrumentations,
      tracerProvider: provider
    });
  }
};
var shouldSendTraces = () => {
  if (!_configuration) {
    return true;
  }
  if (_configuration.traceContent === false || (process.env.TRACELOOP_TRACE_CONTENT || "true").toLowerCase() === "false") {
    return false;
  }
  return true;
};
var forceFlush = async () => {
  await _spanProcessor?.forceFlush();
};

// src/sdk/configuration/index.ts
var _configuration;
var initializeTracing = (options) => {
  if (_configuration) {
    return;
  }
  if (!options.baseUrl) {
    options.baseUrl = process.env.LMNR_BASE_URL || "https://api.lmnr.ai:8443";
  }
  if (!options.apiKey) {
    options.apiKey = process.env.LMNR_PROJECT_API_KEY;
  }
  if (options.apiKey && typeof options.apiKey !== "string") {
    throw new InitializationError('"apiKey" must be a string');
  }
  if (!options.appName) {
    options.appName = process.env.npm_package_name;
  }
  _configuration = Object.freeze(options);
  if (options.logLevel) {
    import_api9.diag.setLogger(
      new import_api9.DiagConsoleLogger(),
      logLevelToOtelLogLevel(options.logLevel)
    );
  }
  startTracing(_configuration);
};
var logLevelToOtelLogLevel = (logLevel) => {
  switch (logLevel) {
    case "debug":
      return import_api9.DiagLogLevel.DEBUG;
    case "info":
      return import_api9.DiagLogLevel.INFO;
    case "warn":
      return import_api9.DiagLogLevel.WARN;
    case "error":
      return import_api9.DiagLogLevel.ERROR;
  }
};

// src/sdk/tracing/decorators.ts
var import_api10 = require("@opentelemetry/api");
var import_core2 = require("@opentelemetry/core");
var import_pino8 = __toESM(require("pino"));
var import_pino_pretty8 = __toESM(require("pino-pretty"));
var logger8 = (0, import_pino8.default)((0, import_pino_pretty8.default)({
  colorize: true,
  minimumLevel: "info"
}));
function observeBase({
  name,
  associationProperties,
  input,
  ignoreInput,
  ignoreOutput,
  inputParameters,
  suppressTracing: shouldSuppressTracing,
  parentSpanContext
}, fn, thisArg, ...args) {
  let entityContext = import_api10.context.active();
  const currentAssociationProperties = entityContext.getValue(ASSOCIATION_PROPERTIES_KEY);
  if (associationProperties) {
    entityContext = entityContext.setValue(
      ASSOCIATION_PROPERTIES_KEY,
      { ...currentAssociationProperties ?? {}, ...associationProperties }
    );
  }
  if (parentSpanContext) {
    const spanContext = tryToOtelSpanContext(parentSpanContext);
    entityContext = import_api10.trace.setSpan(entityContext, import_api10.trace.wrapSpanContext(spanContext));
  }
  if (shouldSuppressTracing) {
    entityContext = (0, import_core2.suppressTracing)(entityContext);
  }
  return import_api10.context.with(
    entityContext,
    () => getTracer().startActiveSpan(
      name,
      {},
      entityContext,
      async (span) => {
        if (shouldSendTraces() && !ignoreInput) {
          try {
            const spanInput = inputParameters ?? args;
            if (input !== void 0) {
              span.setAttribute(
                SPAN_INPUT,
                serialize(input)
              );
            } else if (spanInput.length === 1 && typeof spanInput[0] === "object" && !(spanInput[0] instanceof Map)) {
              span.setAttribute(
                SPAN_INPUT,
                serialize(spanInput[0])
              );
            } else {
              span.setAttribute(
                SPAN_INPUT,
                serialize(spanInput.length > 0 ? spanInput : {})
              );
            }
          } catch (error) {
            logger8.warn(`Failed to serialize input: ${error instanceof Error ? error.message : String(error)}`);
          }
        }
        const { "span_type": spanType, ...rest } = associationProperties ?? {};
        entityContext = entityContext.setValue(
          ASSOCIATION_PROPERTIES_KEY,
          { ...currentAssociationProperties ?? {}, ...rest }
        );
        let res;
        try {
          res = fn.apply(thisArg, args);
        } catch (error) {
          span.recordException(error);
          span.end();
          throw error;
        }
        if (res instanceof Promise) {
          return res.then((resolvedRes) => {
            try {
              if (shouldSendTraces() && !ignoreOutput) {
                span.setAttribute(
                  SPAN_OUTPUT,
                  serialize(resolvedRes)
                );
              }
            } catch (error) {
              logger8.warn(`Failed to serialize async output: ${error instanceof Error ? error.message : String(error)}`);
            } finally {
              span.end();
            }
            return resolvedRes;
          }).catch((error) => {
            span.recordException(error);
            span.end();
            throw error;
          });
        }
        try {
          if (shouldSendTraces() && !ignoreOutput) {
            span.setAttribute(
              SPAN_OUTPUT,
              serialize(res)
            );
          }
        } catch (error) {
          logger8.warn(`Failed to serialize output: ${error instanceof Error ? error.message : String(error)}`);
        } finally {
          span.end();
        }
        return res;
      }
    )
  );
}
var normalizePayload = (payload, seen) => {
  if (payload instanceof Map) {
    return Array.from(payload.entries());
  } else if (Array.isArray(payload)) {
    return payload.map((value) => normalizePayload(value, seen));
  } else if (!payload) {
    return payload;
  } else if (typeof payload === "object") {
    if (seen.has(payload)) {
      return "[Circular reference]";
    }
    seen.add(payload);
    const output = {};
    Object.entries(payload).forEach(([key, value]) => {
      output[key] = normalizePayload(value, seen);
    });
    return output;
  }
  return payload;
};
var serialize = (payload) => JSON.stringify(normalizePayload(payload, /* @__PURE__ */ new WeakSet()));

// src/decorators.ts
async function observe({
  name,
  sessionId,
  traceType,
  spanType,
  input,
  ignoreInput,
  ignoreOutput,
  parentSpanContext
}, fn, ...args) {
  if (fn === void 0 || typeof fn !== "function") {
    throw new Error("Invalid `observe` usage. Second argument `fn` must be a function.");
  }
  let associationProperties = {};
  if (sessionId) {
    associationProperties = { ...associationProperties, "session_id": sessionId };
  }
  if (traceType) {
    associationProperties = { ...associationProperties, "trace_type": traceType };
  }
  if (spanType) {
    associationProperties = { ...associationProperties, "span_type": spanType };
  }
  return await observeBase({
    name: name ?? fn.name,
    associationProperties,
    input,
    ignoreInput,
    ignoreOutput,
    parentSpanContext
  }, fn, void 0, ...args);
}
function withLabels(labels, fn, ...args) {
  let entityContext = import_api11.context.active();
  const currentAssociationProperties = entityContext.getValue(ASSOCIATION_PROPERTIES_KEY) ?? {};
  const oldLabels = currentAssociationProperties.labels ?? [];
  const newLabels = [...oldLabels, ...labels];
  entityContext = entityContext.setValue(
    ASSOCIATION_PROPERTIES_KEY,
    {
      ...currentAssociationProperties,
      labels: newLabels
    }
  );
  const result = import_api11.context.with(entityContext, () => fn(...args));
  const newAssociationProperties = entityContext.getValue(ASSOCIATION_PROPERTIES_KEY) ?? {};
  entityContext = entityContext.setValue(
    ASSOCIATION_PROPERTIES_KEY,
    { ...newAssociationProperties, labels: oldLabels }
  );
  return result;
}
function withTracingLevel(tracingLevel, fn, ...args) {
  let entityContext = import_api11.context.active();
  const currentAssociationProperties = entityContext.getValue(ASSOCIATION_PROPERTIES_KEY) ?? {};
  entityContext = entityContext.setValue(
    ASSOCIATION_PROPERTIES_KEY,
    { ...currentAssociationProperties, "tracing_level": tracingLevel }
  );
  const result = import_api11.context.with(entityContext, () => fn(...args));
  const newAssociationProperties = entityContext.getValue(ASSOCIATION_PROPERTIES_KEY) ?? {};
  delete newAssociationProperties["tracing_level"];
  entityContext = entityContext.setValue(
    ASSOCIATION_PROPERTIES_KEY,
    newAssociationProperties
  );
  return result;
}

// src/evaluations.ts
var import_api12 = require("@opentelemetry/api");
var import_cli_progress = __toESM(require("cli-progress"));
var import_pino9 = __toESM(require("pino"));
var import_pino_pretty9 = __toESM(require("pino-pretty"));
var DEFAULT_CONCURRENCY = 5;
var MAX_EXPORT_BATCH_SIZE = 64;
var logger9 = (0, import_pino9.default)((0, import_pino_pretty9.default)({
  colorize: true,
  minimumLevel: "info"
}));
var getEvaluationUrl = (projectId, evaluationId, baseUrl) => {
  let url = baseUrl ?? "https://api.lmnr.ai";
  if (url === "https://api.lmnr.ai") {
    url = "https://www.lmnr.ai";
  }
  url = url.replace(/\/$/, "");
  if (url.endsWith("localhost") || url.endsWith("127.0.0.1")) {
    url = url + ":5667";
  }
  return `${url}/project/${projectId}/evaluations/${evaluationId}`;
};
var getAverageScores = (results) => {
  const perScoreValues = {};
  for (const result of results) {
    for (const key in result.scores) {
      if (perScoreValues[key]) {
        perScoreValues[key].push(result.scores[key]);
      } else {
        perScoreValues[key] = [result.scores[key]];
      }
    }
  }
  const averageScores = {};
  for (const key in perScoreValues) {
    averageScores[key] = perScoreValues[key].reduce((a, b) => a + b, 0) / perScoreValues[key].length;
  }
  return averageScores;
};
var HumanEvaluator = class {
  constructor(queueName) {
    this.queueName = queueName;
  }
};
var EvaluationReporter = class {
  constructor() {
    this.cliProgress = new import_cli_progress.default.SingleBar(
      {},
      import_cli_progress.default.Presets.shades_classic
    );
    this.progressCounter = 0;
  }
  start({ length }) {
    this.cliProgress.start(length, 0);
  }
  update(batchLength) {
    this.progressCounter += batchLength;
    this.cliProgress.update(this.progressCounter);
  }
  // Call either error or stop, not both
  stopWithError(error) {
    this.cliProgress.stop();
    process.stdout.write(`
Error: ${error.message}
`);
  }
  // Call either error or stop, not both
  stop({
    averageScores,
    projectId,
    evaluationId
  }) {
    this.cliProgress.stop();
    process.stdout.write(`
Check results at ${getEvaluationUrl(projectId, evaluationId)}
`);
    process.stdout.write("\nAverage scores:\n");
    for (const key in averageScores) {
      process.stdout.write(`${key}: ${JSON.stringify(averageScores[key])}
`);
    }
    process.stdout.write("\n");
  }
};
var Evaluation = class {
  constructor({
    data,
    executor,
    evaluators,
    humanEvaluators,
    groupName,
    name,
    config
  }) {
    this.isFinished = false;
    this.concurrencyLimit = DEFAULT_CONCURRENCY;
    this.traceDisableBatch = false;
    this.traceExportBatchSize = MAX_EXPORT_BATCH_SIZE;
    this.uploadPromises = [];
    if (Object.keys(evaluators).length === 0) {
      throw new Error("No evaluators provided");
    }
    const evaluatorNameRegex = /^[\w\s-]+$/;
    for (const key2 in evaluators) {
      if (!evaluatorNameRegex.test(key2)) {
        throw new Error(
          `Invalid evaluator key: "${key2}".Keys must only contain letters, digits, hyphens, underscores, or spaces.`
        );
      }
    }
    this.progressReporter = new EvaluationReporter();
    this.data = data;
    this.executor = executor;
    this.evaluators = evaluators;
    this.humanEvaluators = humanEvaluators;
    this.groupName = groupName;
    this.name = name;
    const key = config?.projectApiKey ?? process.env.LMNR_PROJECT_API_KEY;
    if (key === void 0) {
      throw new Error(
        "Please initialize the Laminar object with your project API key or set the LMNR_PROJECT_API_KEY environment variable"
      );
    }
    const url = config?.baseUrl ?? process?.env?.LMNR_BASE_URL ?? "https://api.lmnr.ai";
    const httpPort = config?.httpPort ?? (url.match(/:\d{1,5}$/g) ? parseInt(url.match(/:\d{1,5}$/g)[0].slice(1)) : 443);
    const urlWithoutSlash = url.replace(/\/$/, "").replace(/:\d{1,5}$/g, "");
    const baseHttpUrl = `${urlWithoutSlash}:${httpPort}`;
    this.client = new LaminarClient({
      baseUrl: baseHttpUrl,
      projectApiKey: key
    });
    if (config) {
      if (config.concurrencyLimit && config.concurrencyLimit < 1) {
        logger9.warn(
          `concurrencyLimit must be greater than 0. Setting to default of ${DEFAULT_CONCURRENCY}`
        );
        this.concurrencyLimit = DEFAULT_CONCURRENCY;
      } else {
        this.concurrencyLimit = config.concurrencyLimit ?? DEFAULT_CONCURRENCY;
      }
      this.traceDisableBatch = config.traceDisableBatch ?? false;
      this.traceExportTimeoutMillis = config.traceExportTimeoutMillis;
      this.traceExportBatchSize = config.traceExportBatchSize ?? MAX_EXPORT_BATCH_SIZE;
    }
    if (Laminar.initialized()) {
      return;
    }
    Laminar.initialize({
      projectApiKey: config?.projectApiKey,
      baseUrl: config?.baseUrl,
      httpPort: config?.httpPort,
      grpcPort: config?.grpcPort,
      instrumentModules: config?.instrumentModules,
      disableBatch: this.traceDisableBatch,
      traceExportTimeoutMillis: this.traceExportTimeoutMillis,
      maxExportBatchSize: this.traceExportBatchSize
    });
  }
  async run() {
    if (this.isFinished) {
      throw new Error("Evaluation is already finished");
    }
    this.progressReporter.start({ length: await this.getLength() });
    let resultDatapoints;
    if (this.data instanceof LaminarDataset) {
      this.data.setClient(this.client);
    }
    try {
      const evaluation = await this.client.evals.init(this.name, this.groupName);
      resultDatapoints = await this.evaluateInBatches(evaluation.id);
      const averageScores = getAverageScores(resultDatapoints);
      if (this.uploadPromises.length > 0) {
        await Promise.all(this.uploadPromises);
      }
      this.progressReporter.stop({
        averageScores,
        projectId: evaluation.projectId,
        evaluationId: evaluation.id
      });
      this.isFinished = true;
      await Laminar.shutdown();
    } catch (e) {
      this.progressReporter.stopWithError(e);
      this.isFinished = true;
      return;
    }
  }
  async evaluateInBatches(evalId) {
    const semaphore = new Semaphore(this.concurrencyLimit);
    const tasks = [];
    const evaluateTask = async (datapoint, index) => {
      try {
        const result = await this.evaluateDatapoint(evalId, datapoint, index);
        this.progressReporter.update(1);
        return [index, result];
      } finally {
        semaphore.release();
      }
    };
    for (let i = 0; i < await this.getLength(); i++) {
      await semaphore.acquire();
      const datapoint = Array.isArray(this.data) ? this.data[i] : await this.data.get(i);
      tasks.push(evaluateTask(datapoint, i));
    }
    const results = await Promise.all(tasks);
    return results.sort((a, b) => a[0] - b[0]).map(([, result]) => result);
  }
  async evaluateDatapoint(evalId, datapoint, index) {
    return observe({ name: "evaluation", traceType: "EVALUATION" }, async () => {
      import_api12.trace.getActiveSpan().setAttribute(SPAN_TYPE, "EVALUATION");
      const executorSpan = Laminar.startSpan({
        name: "executor",
        input: datapoint.data
      });
      executorSpan.setAttribute(SPAN_TYPE, "EXECUTOR");
      const executorSpanId = otelSpanIdToUUID(executorSpan.spanContext().spanId);
      const datapointId = newUUID();
      const partialDatapoint = {
        id: datapointId,
        data: datapoint.data,
        traceId: otelTraceIdToUUID(import_api12.trace.getActiveSpan().spanContext().traceId),
        // For now, all human evaluators are added to every datapoint
        // In the future, we will allow to specify which evaluators are
        // added to a particular datapoint, e.g. random sampling.
        humanEvaluators: this.humanEvaluators,
        executorSpanId,
        index
      };
      await this.client.evals.saveDatapoints({
        evalId,
        datapoints: [partialDatapoint],
        groupName: this.groupName
      });
      const output = await Laminar.withSpan(
        executorSpan,
        async () => await this.executor(datapoint.data)
      );
      const target = datapoint.target;
      let scores = {};
      for (const [evaluatorName, evaluator] of Object.entries(this.evaluators)) {
        const value = await observe(
          { name: evaluatorName },
          async (output2, target2) => {
            import_api12.trace.getActiveSpan().setAttribute(SPAN_TYPE, "EVALUATOR");
            return await evaluator(output2, target2);
          },
          output,
          target
        );
        if (typeof value === "number") {
          if (isNaN(value)) {
            throw new Error(`Evaluator ${evaluatorName} returned NaN`);
          }
          scores[evaluatorName] = value;
        } else {
          scores = { ...scores, ...value };
        }
      }
      const resultDatapoint = {
        id: datapointId,
        executorOutput: output,
        data: datapoint.data,
        target,
        scores,
        traceId: otelTraceIdToUUID(import_api12.trace.getActiveSpan().spanContext().traceId),
        // For now, all human evaluators are added to every datapoint
        // In the future, we will allow to specify which evaluators are
        // added to a particular datapoint, e.g. random sampling.
        humanEvaluators: this.humanEvaluators,
        executorSpanId,
        index
      };
      const uploadPromise = this.client.evals.saveDatapoints({
        evalId,
        datapoints: [resultDatapoint],
        groupName: this.groupName
      });
      this.uploadPromises.push(uploadPromise);
      return resultDatapoint;
    });
  }
  async getLength() {
    return this.data instanceof EvaluationDataset ? await this.data.size() : this.data.length;
  }
};
async function evaluate({
  data,
  executor,
  evaluators,
  humanEvaluators,
  groupName,
  name,
  config
}) {
  const evaluation = new Evaluation({
    data,
    executor,
    evaluators,
    humanEvaluators,
    name,
    groupName,
    config
  });
  if (globalThis._set_global_evaluation) {
    globalThis._evaluation = evaluation;
  } else {
    await evaluation.run();
  }
}

// src/types.ts
var TracingLevel = /* @__PURE__ */ ((TracingLevel2) => {
  TracingLevel2["OFF"] = "off";
  TracingLevel2["META_ONLY"] = "meta_only";
  TracingLevel2["ALL"] = "all";
  return TracingLevel2;
})(TracingLevel || {});

// src/index.ts
var import_api13 = require("@opentelemetry/api");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Dataset,
  HumanEvaluator,
  Laminar,
  LaminarAttributes,
  LaminarClient,
  LaminarDataset,
  Span,
  TracingLevel,
  evaluate,
  observe,
  withLabels,
  withTracingLevel
});
